// @matechat/core/Header/index.js
import "./index.css";
import { defineComponent as d, openBlock as n, createElementBlock as r, createElementVNode as s, normalizeClass as m, createCommentVNode as g, toDisplayString as p, renderSlot as _ } from "vue";
const k = {
  logoImg: {
    type: String,
    default: ""
  },
  title: {
    type: String,
    default: ""
  },
  logoClickable: {
    type: Boolean,
    default: !1
  }
}, f = { class: "mc-header" }, h = ["src", "alt"], u = { class: "mc-header-title" }, C = { class: "mc-header-operation" }, v = /* @__PURE__ */ d({
  __name: "Header",
  props: k,
  emits: ["logoClicked"],
  setup(e, { emit: l }) {
    const t = l, c = e, a = () => {
      c.logoClickable && t("logoClicked");
    };
    return (o, I) => (n(), r("div", f, [
      s(
        "div",
        {
          class: m(["mc-header-logo-container", { clickable: o.logoClickable }]),
          onClick: a
        },
        [
          o.logoImg ? (n(), r("img", {
            key: 0,
            class: "mc-header-logo",
            src: o.logoImg,
            alt: o.logoImg
          }, null, 8, h)) : g("v-if", !0),
          s(
            "div",
            u,
            p(o.title),
            1
            /* TEXT */
          )
        ],
        2
        /* CLASS */
      ),
      s("div", C, [
        _(o.$slots, "operationArea", {}, void 0, !0)
      ])
    ]));
  }
}), y = (e, l) => {
  const t = e.__vccOpts || e;
  for (const [c, a] of l)
    t[c] = a;
  return t;
}, i = /* @__PURE__ */ y(v, [["__scopeId", "data-v-a864dc89"]]);
i.install = (e) => {
  e.component("McHeader", i);
};
export {
  i as McHeader
};
