// @matechat/core/Layout/index.js
import "./index.css";
import { openBlock as n, createElementBlock as c, renderSlot as o } from "vue";
const s = (e, t) => {
  const r = e.__vccOpts || e;
  for (const [a, _] of t)
    r[a] = _;
  return r;
}, u = {}, f = { class: "mc-layout-aside" };
function i(e, t) {
  return n(), c("div", f, [
    o(e.$slots, "default")
  ]);
}
const l = /* @__PURE__ */ s(u, [["render", i]]), m = {}, $ = { class: "mc-layout-content" };
function y(e, t) {
  return n(), c("div", $, [
    o(e.$slots, "default", {}, void 0, !0)
  ]);
}
const h = /* @__PURE__ */ s(m, [["render", y], ["__scopeId", "data-v-629ef113"]]), v = {}, L = { class: "mc-layout-header" };
function M(e, t) {
  return n(), c("div", L, [
    o(e.$slots, "default", {}, void 0, !0)
  ]);
}
const p = /* @__PURE__ */ s(v, [["render", M], ["__scopeId", "data-v-8201933b"]]), k = {}, I = { class: "mc-layout" };
function S(e, t) {
  return n(), c("div", I, [
    o(e.$slots, "default", {}, void 0, !0)
  ]);
}
const d = /* @__PURE__ */ s(k, [["render", S], ["__scopeId", "data-v-8f271543"]]), x = {}, A = { class: "mc-layout-sender" };
function B(e, t) {
  return n(), c("div", A, [
    o(e.$slots, "default")
  ]);
}
const C = /* @__PURE__ */ s(x, [["render", B]]);
d.install = (e) => {
  e.component("McLayoutAside", l), e.component("McLayoutContent", h), e.component("McLayoutHeader", p), e.component("McLayout", d), e.component("McLayoutSender", C);
};
export {
  d as McLayout,
  l as McLayoutAside,
  h as McLayoutContent,
  p as McLayoutHeader,
  C as McLayoutSender
};
