// @matechat/core/Attachment/index.js
import "./index.css";
import { computed as U, defineComponent as E, inject as N, ref as b, watch as S, onMounted as R, onBeforeUnmount as V, openBlock as T, createBlock as k, Teleport as M, createElementVNode as w, normalizeClass as I, renderSlot as A, useSlots as O, useModel as B, provide as $, createElementBlock as j, unref as g, withCtx as z, createTextVNode as _, toDisplayString as H, createCommentVNode as L, mergeModels as x } from "vue";
import { useMcI18n as W } from "@matechat/core/Locale";
const q = {
  uploadOptions: {
    type: Object,
    default: () => ({})
  },
  disabled: {
    type: Boolean,
    default: !1
  },
  accept: {
    type: String,
    default: ""
  },
  dropPlaceholder: {
    type: String
  },
  maxCount: {
    type: Number,
    default: Number.POSITIVE_INFINITY
  },
  maxSize: {
    type: Number,
    default: Number.POSITIVE_INFINITY
  },
  multiple: {
    type: Boolean,
    default: !0
  },
  draggable: {
    type: Boolean,
    default: !0
  },
  beforeUpload: {
    type: Function,
    default: null
  },
  getDropContainer: {
    type: Function
  }
}, Y = {
  change: (a, s) => a instanceof File && Array.isArray(s),
  success: (a, s, o) => a instanceof File && Array.isArray(o),
  error: (a, s, o) => a instanceof File && Array.isArray(o),
  progress: (a, s) => a instanceof File && Array.isArray(s),
  drop: (a) => Array.isArray(a),
  validChange: (a) => Array.isArray(a)
}, P = "mc-attachment";
function K(a) {
  const { file: s, options: o, onProgress: i, onSuccess: h, onError: p } = a, t = new XMLHttpRequest();
  t.upload.onprogress = (l) => {
    if (l.lengthComputable) {
      const r = Math.round(l.loaded * 100 / l.total);
      i(r);
    }
  }, t.onerror = () => {
    p({
      message: "Upload failed due to a network error.",
      status: t.status,
      statusText: t.statusText
    });
  }, t.onload = () => {
    if (t.status >= 200 && t.status < 300) {
      const l = o.responseType === "text" || !t.response ? t.responseText : t.response;
      h(l);
    } else
      p({
        message: `Upload failed with status: ${t.status}`,
        status: t.status,
        statusText: t.statusText,
        response: t.response
      });
  };
  const c = new FormData(), m = o.fileFieldName || "file";
  if (c.append(m, s, s.name), o.additionalParameter)
    for (const l of Object.keys(o.additionalParameter))
      c.append(l, o.additionalParameter[l]);
  if (t.open(o.method || "POST", o.uri, !0), o.withCredentials && (t.withCredentials = !0), o.responseType && (t.responseType = o.responseType), o.headers)
    for (const l of Object.keys(o.headers))
      t.setRequestHeader(l, o.headers[l]);
  if (o.authToken) {
    const l = o.authTokenHeader || "Authorization";
    t.setRequestHeader(l, o.authToken);
  }
  t.send(c);
}
let X = Date.now();
function G(a, s, o, i) {
  const h = U(() => a.disabled || i.value.length >= a.maxCount), p = (r) => {
    const d = URL.createObjectURL(r);
    return {
      uid: X++,
      name: r.name,
      size: r.size,
      type: r.type,
      status: "uploading",
      percentage: 0,
      url: d,
      thumbUrl: d
    };
  }, t = async (r) => {
    if (r.length === 0) return;
    if (i.value.length + r.length > a.maxCount) {
      s("validChange", [{ type: "exceedCount" }]);
      return;
    }
    const d = [], u = [];
    for (const n of r) {
      let e = !0;
      if (a.accept && a.accept !== "*") {
        const f = a.accept.split(",").map((v) => v.trim().toLowerCase()), y = n.type.toLowerCase(), C = n.name.toLowerCase();
        f.some((v) => v.endsWith("/*") ? y.startsWith(v.slice(0, -1)) : v.startsWith(".") ? C.endsWith(v) : y === v) || (u.push({ type: "unsupportedFileType", file: n }), e = !1);
      }
      if (e && n.size / 1024 / 1024 > a.maxSize && (u.push({ type: "exceedSizeLimit", file: n }), e = !1), e && a.beforeUpload)
        try {
          await Promise.resolve(a.beforeUpload(n)) === !1 && (u.push({ type: "beforeUploadRejected", file: n }), e = !1);
        } catch {
          u.push({ type: "beforeUploadRejected", file: n }), e = !1;
        }
      e && d.push(n);
    }
    if (u.length > 0 && s("validChange", u), d.length !== 0)
      for (const n of d) {
        const e = p(n);
        i.value.push(e), s("change", n, [...i.value]), c(n, e);
      }
  }, c = (r, d) => {
    const u = () => i.value.findIndex((n) => n.uid === d.uid);
    K({
      file: r,
      options: a.uploadOptions,
      onProgress: (n) => {
        const e = u();
        e > -1 && (i.value[e].percentage = n, s("progress", r, [...i.value]));
      },
      onSuccess: (n) => {
        const e = u();
        e > -1 && (i.value[e].status = "success", i.value[e].response = n, s("success", r, n, [...i.value]));
      },
      onError: (n) => {
        const e = u();
        e > -1 && (i.value[e].status = "uploadError", i.value[e].error = n, s("error", r, n, [...i.value]));
      }
    });
  };
  return {
    isDisabled: h,
    handleClick: () => {
      var r;
      h.value || (r = o.value) == null || r.click();
    },
    handleFileChange: (r) => {
      const d = r.target.files;
      d && t(Array.from(d));
    },
    uploadFiles: t
  };
}
const J = /* @__PURE__ */ E({
  __name: "drop-area",
  setup(a) {
    const { rootProps: s, rootEmits: o, isDisabled: i, uploadFiles: h } = N(
      P
    ), p = b(), t = b(), c = b(!1);
    let m = 0;
    S(
      () => s.getDropContainer,
      (e) => {
        if (e) {
          const f = e();
          p.value !== f && (p.value = f);
        }
      }
    );
    const l = (e) => {
      if (e.preventDefault(), m++, m === 1 && (c.value = !0), p.value && t.value)
        if (p.value === document.body)
          t.value.style.inset = "0";
        else {
          const { x: f, y, width: C, height: D } = p.value.getBoundingClientRect();
          t.value.style.width = `${C}px`, t.value.style.height = `${D}px`, t.value.style.top = `${y}px`, t.value.style.left = `${f}px`;
        }
    }, r = (e) => {
      e.preventDefault();
    }, d = (e) => {
      e.preventDefault(), m--, m === 0 && (c.value = !1);
    }, u = (e) => {
      var y;
      if (e.preventDefault(), c.value = !1, m = 0, i.value) return;
      const f = Array.from(((y = e.dataTransfer) == null ? void 0 : y.files) || []);
      f.length > 0 && (o("drop", f), h(f));
    };
    function n(e) {
      e.preventDefault(), c.value = !1, m = 0;
    }
    return R(() => {
      s.getDropContainer && (p.value = s.getDropContainer()), document.body.addEventListener("dragenter", l), document.body.addEventListener("dragover", r), document.body.addEventListener("dragleave", d), document.body.addEventListener("drop", n);
    }), V(() => {
      document.body.removeEventListener("dragenter", l), document.body.removeEventListener("dragover", r), document.body.removeEventListener("dragleave", d), document.body.removeEventListener("drop", n);
    }), (e, f) => (T(), k(M, { to: "body" }, [
      w(
        "div",
        {
          ref_key: "dropAreaEl",
          ref: t,
          class: I(["mc-attachment-drop-area", !c.value && "hide-drop-area"]),
          onDrop: u
        },
        [
          A(e.$slots, "default")
        ],
        34
        /* CLASS, NEED_HYDRATION */
      )
    ]));
  }
}), Q = ["accept", "multiple", "disabled"], F = /* @__PURE__ */ E({
  name: "McAttachment",
  __name: "Attachment",
  props: /* @__PURE__ */ x(q, {
    modelValue: { default: [] },
    modelModifiers: {}
  }),
  emits: /* @__PURE__ */ x(Y, ["update:modelValue"]),
  setup(a, { emit: s }) {
    const o = a, i = s;
    O();
    const h = B(a, "modelValue"), { t: p } = W(), t = b(), { handleClick: c, handleFileChange: m, isDisabled: l, uploadFiles: r } = G(
      o,
      i,
      t,
      h
    );
    return $(P, {
      rootProps: o,
      rootEmits: i,
      isDisabled: l,
      uploadFiles: r
    }), (d, u) => (T(), j(
      "div",
      {
        class: I(["mc-attachment", { "is-disabled": g(l) }]),
        onClick: u[1] || (u[1] = //@ts-ignore
        (...n) => g(c) && g(c)(...n))
      },
      [
        A(d.$slots, "default"),
        w("input", {
          ref_key: "inputRef",
          ref: t,
          type: "file",
          class: "mc-attachment-file",
          accept: d.accept,
          multiple: d.multiple,
          disabled: g(l),
          onChange: u[0] || (u[0] = //@ts-ignore
          (...n) => g(m) && g(m)(...n))
        }, null, 40, Q),
        d.draggable ? (T(), k(J, { key: 0 }, {
          default: z(() => [
            A(d.$slots, "dropPlaceholder", {}, () => [
              _(
                H(d.dropPlaceholder ?? g(p)("Attachment.dragToUpload")),
                1
                /* TEXT */
              )
            ])
          ]),
          _: 3
          /* FORWARDED */
        })) : L("v-if", !0)
      ],
      2
      /* CLASS */
    ));
  }
});
F.install = (a) => {
  a.component("McAttachment", F);
};
export {
  F as McAttachment
};
