// @matechat/core/Input/index.js
import "./index.css";
import { ref as w, defineComponent as X, inject as oe, computed as ae, watch as Y, onMounted as te, withDirectives as be, openBlock as m, createElementBlock as x, normalizeClass as q, unref as g, normalizeStyle as ce, isRef as ye, vModelText as xe, nextTick as de, createElementVNode as $, withModifiers as le, toDisplayString as Z, resolveComponent as Ce, createCommentVNode as _, createBlock as O, withCtx as pe, renderSlot as A, Fragment as ee, renderList as Te, createTextVNode as Le, reactive as Se, provide as _e, createVNode as ie } from "vue";
import { useMcI18n as fe } from "@matechat/core/Locale";
var V = /* @__PURE__ */ ((n) => (n.Simple = "simple", n.Full = "full", n))(V || {}), me = /* @__PURE__ */ ((n) => (n.Bordered = "bordered", n.BorderLess = "borderless", n))(me || {}), ne = /* @__PURE__ */ ((n) => (n.Simple = "simple", n.Full = "full", n))(ne || {}), K = /* @__PURE__ */ ((n) => (n.Enter = "enter", n.ShiftEnter = "shiftEnter", n))(K || {});
const we = {
  minRows: 1,
  maxRows: 5
}, ke = {
  value: {
    type: String,
    default: ""
  },
  formatContentOptions: {
    type: Object,
    default: () => {
    }
  },
  placeholder: {
    type: String
  },
  disabled: {
    type: Boolean,
    default: !1
  },
  displayType: {
    type: String,
    default: "full"
    /* Full */
  },
  variant: {
    type: String,
    default: "bordered"
    /* Bordered */
  },
  sendBtnVariant: {
    type: String,
    default: "full"
    /* Full */
  },
  loading: {
    type: Boolean,
    default: !1
  },
  showCount: {
    type: Boolean,
    default: !1
  },
  maxLength: {
    type: Number
  },
  submitShortKey: {
    type: [String, null],
    default: "enter"
    /* Enter */
  },
  autofocus: {
    type: Boolean,
    default: !1
  },
  autosize: {
    type: [Boolean, Object],
    default: !1
  },
  autoClear: {
    type: Boolean,
    default: !0
  }
}, Ee = ["change", "submit", "cancel", "focus", "blur"], Q = "mc-input", Ie = [
  "letter-spacing",
  "line-height",
  "padding-top",
  "padding-bottom",
  "font-family",
  "font-weight",
  "font-size",
  "font-variant",
  "text-rendering",
  "text-transform",
  "width",
  "text-indent",
  "padding-left",
  "padding-right",
  "border-width",
  "box-sizing",
  "word-break",
  "white-space"
], Be = `
  min-height:0 !important;
  max-height:none !important;
  height:0 !important;
  visibility:hidden !important;
  overflow:hidden !important;
  position:absolute !important;
  z-index:-1000 !important;
  top:0 !important;
  right:0 !important;
  pointer-events: none !important;
`;
function $e(n) {
  const a = window.getComputedStyle(n), s = a.getPropertyValue("box-sizing"), t = Number.parseFloat(a.getPropertyValue("padding-top")) + Number.parseFloat(a.getPropertyValue("padding-bottom")), l = Number.parseFloat(a.getPropertyValue("border-top-width")) + Number.parseFloat(a.getPropertyValue("border-bottom-width"));
  return { contextStyle: Ie.map(
    (r) => `${r}:${a.getPropertyValue(r)}`
  ).join(";"), paddingSize: t, borderSize: l, boxSizing: s };
}
function Ne(n, a = 1, s) {
  const { contextStyle: t, paddingSize: l, borderSize: i, boxSizing: r } = $e(n), u = document.createElement("textarea");
  u.setAttribute("style", `${t};${Be}`), u.value = n.value || n.placeholder || "", document.body.appendChild(u);
  let f = u.scrollHeight;
  const h = {
    height: `${f}px`
  };
  if (a !== void 0 || s !== void 0) {
    u.value = " ";
    const C = u.scrollHeight - l;
    if (a !== void 0) {
      let d = C * a;
      r === "border-box" && (d = d + l + i), f = Math.max(d, f), h.minHeight = `${d}px`;
    }
    if (s !== void 0) {
      let d = C * s;
      r === "border-box" && (d = d + l + i), f > d ? (f = d, h.overflowY = "auto") : h.overflowY = "hidden";
    }
  }
  return h.height = `${f}px`, u.remove(), h;
}
function Me(n) {
  const { textareaRef: a, autosize: s } = n, t = w({}), l = () => typeof s == "boolean" ? we : s;
  return {
    textareaStyle: t,
    updateTextareaStyle: () => {
      if (!a.value || s === !1) {
        t.value = {};
        return;
      }
      const r = l(), u = Ne(
        a.value,
        r.minRows,
        r.maxRows
      );
      t.value = {
        ...u,
        resize: "none"
      };
    }
  };
}
const De = ["placeholder", "disabled", "maxlength"], Re = /* @__PURE__ */ X({
  __name: "textarea",
  setup(n) {
    const { t: a } = fe(), s = w(null), { inputValue: t, rootProps: l, rootEmits: i, clearInputAfterSubmit: r } = oe(
      Q
    ), u = ae(() => {
      let v = "", E = "";
      return l.submitShortKey === K.Enter && (v = "Enter", E = "Shift + Enter"), l.submitShortKey === K.ShiftEnter && (v = "Shift + Enter", E = "Enter"), l.placeholder ?? (v ? a("Input.pleaseEnterPlaceholder", {
        enterKey: v,
        shiftEnterKey: E
      }) : a("Input.pleaseEnter"));
    }), { textareaStyle: f, updateTextareaStyle: h } = Me({
      textareaRef: s,
      autosize: l.autosize
    });
    Y(
      () => t.value,
      () => {
        h();
      }
    ), te(() => {
      h();
    });
    let C = !1;
    const d = () => {
      de(() => {
        i("change", t.value);
      });
    }, T = () => {
      C || d();
    }, b = () => {
      C = !0;
    }, B = () => {
      setTimeout(() => {
        C = !1, d();
      }, 10);
    }, z = (v) => {
      if (l.submitShortKey === null)
        return;
      (l.submitShortKey === K.Enter ? !v.shiftKey : l.submitShortKey === K.ShiftEnter ? v.shiftKey : !1) && v.key === "Enter" && !C && (v.preventDefault(), i("submit", t.value), r(), i("change", t.value));
    }, F = (v) => {
      i("focus", v);
    }, k = (v) => {
      i("blur", v);
    };
    return te(() => {
      l.autofocus && s.value && !l.disabled && s.value.focus();
    }), (v, E) => be((m(), x("textarea", {
      ref_key: "textareaRef",
      ref: s,
      "onUpdate:modelValue": E[0] || (E[0] = (H) => ye(t) ? t.value = H : null),
      placeholder: u.value,
      disabled: g(l).disabled,
      maxlength: g(l).maxLength,
      style: ce(g(f)),
      class: q([
        "mc-textarea",
        { "mc-textarea-simple": g(l).displayType === g(V).Simple, "mc-textarea-disabled": g(l).disabled }
      ]),
      onInput: T,
      onCompositionstart: b,
      onCompositionend: B,
      onKeydown: z,
      onFocus: F,
      onBlur: k
    }, null, 46, De)), [
      [xe, g(t)]
    ]);
  }
});
function he(n) {
  const a = window.getSelection();
  if (a && a.rangeCount > 0) {
    const s = a.getRangeAt(0);
    s.deleteContents();
    const t = document.createTextNode(n);
    s.insertNode(t), s.setStartAfter(t), s.setEndAfter(t), a.removeAllRanges(), a.addRange(s);
  }
}
const Ae = (n) => {
  var t;
  n.preventDefault();
  const s = (((t = n.clipboardData) == null ? void 0 : t.getData("text/plain")) || "").replace(/[\r\n]+/g, " ");
  s && he(s);
}, Ke = (n) => {
  if (n) {
    const a = document.createRange(), s = window.getSelection();
    a.selectNodeContents(n), a.collapse(!1), s == null || s.removeAllRanges(), s == null || s.addRange(a);
  }
}, Pe = {
  class: "editable-span",
  contenteditable: "false"
}, ze = ["contenteditable", "placeholder"], Fe = /* @__PURE__ */ X({
  __name: "InputTag",
  props: {
    content: String,
    placeholder: {
      type: String,
      default: ""
    },
    disabled: {
      type: Boolean,
      default: !1
    }
  },
  emits: ["change"],
  setup(n, { emit: a }) {
    const s = a, t = n, l = w(!1), i = w(t.content), r = w(null);
    Y(
      () => t.content,
      (d) => {
        i.value = d;
      }
    );
    const u = () => {
      l.value = !t.disabled;
    }, f = () => {
      l.value = !1;
    }, h = (d) => {
      var b;
      const T = ((b = r.value) == null ? void 0 : b.textContent) || "";
      i.value = T, s("change", T);
    }, C = (d) => {
      Ae(d), h();
    };
    return (d, T) => {
      var b, B;
      return m(), x("span", Pe, [
        $("span", {
          ref_key: "editableSpanRef",
          ref: r,
          contenteditable: l.value,
          placeholder: n.placeholder,
          class: q(["input-custom-item", { "input-custom-placeholder": !((b = i.value) != null && b.length), "inline-text": (B = i.value) == null ? void 0 : B.length }]),
          tabindex: "0",
          onInput: le(h, ["stop"]),
          onFocus: u,
          onBlur: f,
          onPaste: le(C, ["stop"])
        }, Z(n.content), 43, ze)
      ]);
    };
  }
}), W = (n, a) => {
  const s = n.__vccOpts || n;
  for (const [t, l] of a)
    s[t] = l;
  return s;
}, Ve = /* @__PURE__ */ W(Fe, [["__scopeId", "data-v-de0a1464"]]), Ze = ["contenteditable", "placeholder"], Oe = { contenteditable: "false" }, He = {
  id: "ai-input-prefix",
  class: "ai-input-prefix"
}, Ue = /* @__PURE__ */ X({
  __name: "EditableBlock",
  props: {
    templateParts: {
      type: Array,
      default: () => []
    },
    placeholder: {
      type: String,
      default: ""
    },
    maxLength: {
      type: Number,
      default: 2e3
    },
    disabled: {
      type: Boolean,
      default: !1
    },
    submitShortKey: {
      type: [String, null],
      default: K.Enter
    },
    autofocus: {
      type: Boolean,
      default: !1
    }
  },
  emits: [
    "send",
    "input",
    "blur",
    "focus",
    "onBlockKeyArrowUp"
  ],
  setup(n, { expose: a, emit: s }) {
    const { rootProps: t } = oe(Q), l = s, i = n, r = w({
      themeTagText: "",
      clearInput: !1,
      popoverContent: "",
      type: "themeTag"
    }), u = w(null), f = w([...i.templateParts]);
    Y(
      () => i.templateParts,
      (e) => {
        f.value = [...e];
      }
    );
    const h = () => {
      const e = u.value;
      e && (e.innerHTML = "", k());
    }, C = () => {
      var p;
      const e = (p = u.value) == null ? void 0 : p.textContent;
      e.length > i.maxLength || (l("send", e), h());
    }, d = () => {
      var e;
      (e = u.value) == null || e.focus(), Ke(u.value);
    }, T = (e) => {
      l("blur", e);
    }, b = (e) => {
      l("focus", e);
    }, B = w(!1), z = () => {
      B.value = !0;
    }, F = () => {
      B.value = !1, k();
    }, k = () => {
      B.value || E();
    }, v = () => {
      var p, c;
      let e = (p = u.value) == null ? void 0 : p.textContent;
      if (r.value.themeTagText) {
        const y = (c = u.value) == null ? void 0 : c.querySelector(
          ".ai-input-prefix-wrapper"
        );
        if (y) {
          const D = y.textContent || "";
          e = e.replace(D, " ");
        }
      }
      return e;
    }, E = () => {
      setTimeout(() => {
        const e = v();
        l("input", e);
      }, 50);
    }, H = (e) => {
      const p = i.submitShortKey === K.Enter ? !e.shiftKey : i.submitShortKey === K.ShiftEnter ? e.shiftKey : !1;
      if (e.key === "Enter" && p && (e.preventDefault(), C()), ["ArrowLeft", "ArrowRight", "Backspace", "Delete"].includes(e.key)) {
        const c = window.getSelection();
        if (!(c != null && c.rangeCount) || !c.isCollapsed)
          return;
        const y = c.getRangeAt(0), { startContainer: D, startOffset: J } = y, S = e.key === "ArrowLeft" || e.key === "Backspace" ? "previous" : "next", P = L(D, J, S);
        if (P && P.nodeType === Node.ELEMENT_NODE) {
          const N = P.querySelector(".input-custom-item");
          if (!N)
            return;
          if (e.preventDefault(), N.contentEditable = "true", e.key === "ArrowLeft" || e.key === "ArrowRight") {
            const R = document.createRange();
            R.setStart(
              N,
              S === "previous" ? N.childNodes.length : 0
            ), R.collapse(!1), c.removeAllRanges(), c.addRange(R);
          } else if (e.key === "Backspace" || e.key === "Delete")
            if (N.textContent.trim()) {
              const R = document.createRange();
              R.setStart(
                N,
                S === "previous" ? N.childNodes.length : 0
              ), R.collapse(!1), c.removeAllRanges(), c.addRange(R);
            } else
              P.remove();
        }
      }
      e.key === "ArrowUp" && o() && l("onBlockKeyArrowUp", e);
    };
    function o() {
      var y;
      const e = window.getSelection();
      if (!(e != null && e.rangeCount))
        return !0;
      const p = e.getRangeAt(0);
      if (p.startOffset !== 0)
        return !1;
      let c = p.startContainer;
      for (; c && !((y = c.classList) != null && y.contains("editable-container")); ) {
        if (c.previousSibling)
          return !1;
        c = c.parentNode;
      }
      return !0;
    }
    function L(e, p, c) {
      var P;
      if (!e)
        return null;
      let y = e;
      const D = (P = e.classList) == null ? void 0 : P.contains("editable-container");
      if (D) {
        const N = e.childNodes, R = Math.min(p, N.length - 1);
        y = N[R];
      }
      if (!D && c === "previous" && p !== 0 || !D && c === "next" && p !== e.textContent.length)
        return null;
      let J = null, S = y;
      for (; S; )
        if (S = c === "next" ? S.nextSibling : S.previousSibling, !!S && S.nodeType !== Node.COMMENT_NODE && (S.nodeType !== Node.TEXT_NODE || S.nodeType === Node.TEXT_NODE && S.nodeValue.trim() !== "")) {
          J = S;
          break;
        }
      return J;
    }
    const I = (e) => {
      var c;
      e.preventDefault();
      const p = (c = (e.originalEvent || e).clipboardData) == null ? void 0 : c.getData("text/plain");
      he(p), k();
    }, M = (e, p, c) => {
      i.disabled || (f.value.push({
        key: e,
        type: "input",
        placeholder: p,
        content: c
      }), G(), k());
    }, U = (e) => {
      i.disabled || (f.value = [...e], G(), k());
    }, ge = (e) => {
      i.disabled || (f.value.push({
        type: "text",
        content: e
      }), G(), k());
    }, ve = (e, p, c) => {
      i.disabled || (r.value = {
        themeTagText: e,
        clearInput: c || !1,
        popoverContent: p,
        type: "themeTag"
      });
    }, se = () => {
      r.value.themeTagText && r.value.clearInput ? h() : (r.value = {
        themeTagText: "",
        clearInput: !1,
        popoverContent: "",
        type: "themeTag"
      }, E());
    }, G = () => {
      de(() => {
        d();
      });
    };
    return te(() => {
      i.autofocus && u.value && !i.disabled && G();
    }), a({
      clearInput: h,
      getInput: v,
      focusInput: d,
      setMixTags: U,
      setText: ge,
      setInputTag: M,
      openTipTag: ve,
      closeTipTag: se
    }), (e, p) => {
      const c = Ce("d-popover");
      return m(), x("div", {
        ref_key: "editableDivRef",
        ref: u,
        class: q(["editable-container", { "mc-input-disabled": i.disabled, "mc-input-simple": g(t).displayType === g(V).Simple }]),
        contenteditable: !i.disabled,
        spellcheck: "false",
        placeholder: i.placeholder,
        onInput: k,
        onKeydown: H,
        onPaste: I,
        onCompositionstart: z,
        onCompositionend: F,
        onBlur: T,
        onFocus: b
      }, [
        _(" 前置标签 "),
        r.value.themeTagText ? (m(), O(c, {
          key: 0,
          trigger: "hover",
          content: r.value.popoverContent
        }, {
          default: pe(() => [
            $("span", Oe, [
              r.value.themeTagText ? (m(), x("span", {
                key: 0,
                class: "ai-input-prefix-wrapper",
                onClick: se,
                contenteditable: "false"
              }, [
                A(e.$slots, "themeTag", { themeTag: r.value }, () => [
                  $(
                    "span",
                    He,
                    Z(r.value.themeTagText),
                    1
                    /* TEXT */
                  )
                ], !0)
              ])) : _("v-if", !0),
              p[0] || (p[0] = $(
                "span",
                {
                  contenteditable: "false",
                  style: { display: "inline-block", width: "4px" }
                },
                null,
                -1
                /* HOISTED */
              ))
            ])
          ]),
          _: 3
          /* FORWARDED */
        }, 8, ["content"])) : _("v-if", !0),
        (m(!0), x(
          ee,
          null,
          Te(f.value, (y, D) => (m(), x(
            ee,
            null,
            [
              y.type === "text" ? (m(), x(
                ee,
                { key: 0 },
                [
                  Le(
                    Z(y.content),
                    1
                    /* TEXT */
                  )
                ],
                64
                /* STABLE_FRAGMENT */
              )) : y.type === "input" ? (m(), O(Ve, {
                key: `input${D}`,
                content: y.content,
                placeholder: y.placeholder,
                disabled: i.disabled,
                onChange: k
              }, null, 8, ["content", "placeholder", "disabled"])) : _("v-if", !0)
            ],
            64
            /* STABLE_FRAGMENT */
          ))),
          256
          /* UNKEYED_FRAGMENT */
        ))
      ], 42, Ze);
    };
  }
}), je = /* @__PURE__ */ W(Ue, [["__scopeId", "data-v-92145e12"]]), Ye = {}, Xe = {
  width: "16px",
  height: "16px",
  viewBox: "0 0 16 16",
  version: "1.1",
  xmlns: "http://www.w3.org/2000/svg"
};
function qe(n, a) {
  return m(), x("svg", Xe, a[0] || (a[0] = [
    $(
      "g",
      {
        stroke: "none",
        "stroke-width": "1",
        fill: "none",
        "fill-rule": "evenodd"
      },
      [
        $("path", { d: "M8,12 C8.27614237,12 8.5,12.2238576 8.5,12.5 L8.5,14.5 C8.5,14.7761424 8.27614237,15 8,15 C7.72385763,15 7.5,14.7761424 7.5,14.5 L7.5,12.5 C7.5,12.2238576 7.72385763,12 8,12 Z M11.0495421,11.3466838 L12.2251126,12.9647178 C12.3874251,13.1881217 12.3379006,13.5008065 12.1144968,13.663119 C11.8910929,13.8254314 11.5784081,13.775907 11.4160956,13.5525031 L10.2405251,11.9344691 C10.0782127,11.7110652 10.1277371,11.3983804 10.351141,11.236068 C10.5745449,11.0737556 10.8872297,11.12328 11.0495421,11.3466838 Z M5.64885899,11.236068 C5.87226287,11.3983804 5.92178728,11.7110652 5.75947486,11.9344691 L4.58390436,13.5525031 C4.42159194,13.775907 4.10890711,13.8254314 3.88550323,13.663119 C3.66209936,13.5008065 3.61257495,13.1881217 3.77488736,12.9647178 L4.95045787,11.3466838 C5.11277028,11.12328 5.42545512,11.0737556 5.64885899,11.236068 Z M12.4342628,8.91504822 L14.3363759,9.53308221 C14.5990029,9.61841489 14.7427283,9.90049196 14.6573956,10.163119 C14.5720629,10.425746 14.2899859,10.5694714 14.0273589,10.4841387 L12.1252458,9.86610473 C11.8626188,9.78077205 11.7188934,9.49869498 11.8042261,9.23606798 C11.8895588,8.97344097 12.1716358,8.82971553 12.4342628,8.91504822 Z M4.19577393,9.23606798 C4.28110662,9.49869498 4.13738118,9.78077205 3.87475417,9.86610473 L1.97264114,10.4841387 C1.71001414,10.5694714 1.42793707,10.425746 1.34260439,10.163119 C1.2572717,9.90049196 1.40099714,9.61841489 1.66362415,9.53308221 L3.56573718,8.91504822 C3.82836418,8.82971553 4.11044125,8.97344097 4.19577393,9.23606798 Z M14.6573956,5.83688104 C14.7427283,6.09950804 14.5990029,6.38158511 14.3363759,6.46691779 L12.4342628,7.08495178 C12.1716358,7.17028447 11.8895588,7.02655903 11.8042261,6.76393202 C11.7188934,6.50130502 11.8626188,6.21922795 12.1252458,6.13389527 L14.0273589,5.51586128 C14.2899859,5.43052859 14.5720629,5.57425403 14.6573956,5.83688104 Z M1.97264114,5.51586128 L3.87475417,6.13389527 C4.13738118,6.21922795 4.28110662,6.50130502 4.19577393,6.76393202 C4.11044125,7.02655903 3.82836418,7.17028447 3.56573718,7.08495178 L1.66362415,6.46691779 C1.40099714,6.38158511 1.2572717,6.09950804 1.34260439,5.83688104 C1.42793707,5.57425403 1.71001414,5.43052859 1.97264114,5.51586128 Z M12.1144968,2.33688104 C12.3379006,2.49919345 12.3874251,2.81187829 12.2251126,3.03528216 L11.0495421,4.65331615 C10.8872297,4.87672003 10.5745449,4.92624444 10.351141,4.76393202 C10.1277371,4.60161961 10.0782127,4.28893477 10.2405251,4.0655309 L11.4160956,2.44749691 C11.5784081,2.22409304 11.8910929,2.17456862 12.1144968,2.33688104 Z M4.58390436,2.44749691 L5.75947486,4.0655309 C5.92178728,4.28893477 5.87226287,4.60161961 5.64885899,4.76393202 C5.42545512,4.92624444 5.11277028,4.87672003 4.95045787,4.65331615 L3.77488736,3.03528216 C3.61257495,2.81187829 3.66209936,2.49919345 3.88550323,2.33688104 C4.10890711,2.17456862 4.42159194,2.22409304 4.58390436,2.44749691 Z M8,1 C8.27614237,1 8.5,1.22385763 8.5,1.5 L8.5,3.5 C8.5,3.77614237 8.27614237,4 8,4 C7.72385763,4 7.5,3.77614237 7.5,3.5 L7.5,1.5 C7.5,1.22385763 7.72385763,1 8,1 Z" })
      ],
      -1
      /* HOISTED */
    )
  ]));
}
const We = /* @__PURE__ */ W(Ye, [["render", qe]]), Ge = {}, Je = {
  width: "16px",
  height: "16px",
  viewBox: "0 0 16 16",
  version: "1.1",
  xmlns: "http://www.w3.org/2000/svg"
};
function Qe(n, a) {
  return m(), x("svg", Je, a[0] || (a[0] = [
    $(
      "g",
      {
        stroke: "none",
        "stroke-width": "1",
        fill: "none",
        "fill-rule": "evenodd"
      },
      [
        $("path", { d: "M13.5105149,2.56164948 C14.1211149,2.35546948 14.7829149,2.68328948 14.9892149,3.29376948 C15.0349149,3.42925948 15.0552149,3.57159948 15.0495149,3.71368948 L15.0495149,3.71368948 L15.0351149,3.85546948 L13.6575149,12.2717195 C13.5237149,13.0892195 12.7524149,13.6435195 11.9348149,13.5097195 C11.8542149,13.4965195 11.7751149,13.4768195 11.6980149,13.4508195 L11.6980149,13.4508195 L11.5840149,13.4071195 L9.3049349,12.4260195 L8.4482449,13.5422195 C8.1958049,13.8708195 7.8205549,14.0786195 7.4128449,14.1207195 L7.4128449,14.1207195 L7.2585449,14.1286195 C6.4787649,14.1286195 5.8381349,13.5337195 5.7653849,12.7731195 L5.7653849,12.7731195 L5.7585449,12.6286195 L5.7602549,10.5258195 L5.7756349,10.4384195 L5.7944349,10.3824195 L5.8266649,10.3168995 L5.8735349,10.2503695 L5.9177249,10.2039195 L12.6909149,3.89392948 L2.3403349,7.38903948 C2.2329149,7.42529948 2.1418449,7.49676948 2.0808149,7.59063948 L2.0808149,7.59063948 L2.0412649,7.66497948 C1.9440949,7.89044948 2.0258749,8.14788948 2.2236349,8.27917948 L2.2236349,8.27917948 L2.3029749,8.32214948 L4.4799849,9.26177948 C4.7333949,9.37121948 4.8503449,9.66546948 4.7409649,9.91894948 C4.6435549,10.1443495 4.4003949,10.2616595 4.1694349,10.2081295 L4.1694349,10.2081295 L4.0837449,10.1798795 L1.9072249,9.24047948 C1.1462449,8.91284948 0.795166898,8.03039948 1.1225549,7.26953948 C1.2700249,6.92755948 1.5380849,6.65435948 1.8730449,6.50018948 L1.8730449,6.50018948 L2.0205049,6.44158948 Z M6.7578149,11.3299195 L6.7585449,12.6286195 L6.7666013,12.7185103 C6.80894095,12.9517898 7.0131049,13.1286195 7.2585449,13.1286195 C7.3828149,13.1286195 7.5014649,13.0824195 7.5925349,13.0008195 L7.5925349,13.0008195 L7.6550349,12.9332195 L8.3569349,12.0179195 L6.7578149,11.3299195 Z M13.9888149,4.05090948 L7.1599149,10.4129195 L11.9795149,12.4886195 C11.9980149,12.4966195 12.0173149,12.5035195 12.0369149,12.5093195 L12.0369149,12.5093195 L12.0964149,12.5228195 C12.3386149,12.5624195 12.5686149,12.4209195 12.6482149,12.1975195 L12.6482149,12.1975195 L12.6706149,12.1101195 L13.9888149,4.05090948 Z" })
      ],
      -1
      /* HOISTED */
    )
  ]));
}
const et = /* @__PURE__ */ W(Ge, [["render", Qe]]), tt = ["disabled"], nt = { class: "mc-button-content" }, ot = { key: 2 }, at = /* @__PURE__ */ X({
  __name: "button",
  setup(n) {
    const { t: a } = fe(), { inputValue: s, rootProps: t, rootEmits: l, clearInputAfterSubmit: i } = oe(
      Q
    ), r = w(!1), u = w(!1), f = Se({
      top: "0px",
      left: "0px"
    }), h = ae(() => ({
      "mc-button": !0,
      "mc-button-loading": t.loading,
      mousedown: r.value,
      "mc-button-simple": t.sendBtnVariant === ne.Simple
    })), C = (T) => {
      f.left = T.offsetX + "px", f.top = T.offsetY + "px", u.value = !0, setTimeout(() => {
        u.value = !1;
      }, 300);
    }, d = (T) => {
      C(T), t.loading ? l("cancel") : (l("submit", s.value), i(), l("change", s.value));
    };
    return (T, b) => (m(), x("button", {
      disabled: g(t).disabled || !g(t).loading && !g(s),
      class: q(h.value),
      onClick: d,
      onMousedown: b[0] || (b[0] = () => r.value = !0),
      onMouseup: b[1] || (b[1] = () => r.value = !1)
    }, [
      $("span", nt, [
        g(t).loading ? (m(), O(We, { key: 0 })) : (m(), O(et, { key: 1 })),
        g(t).sendBtnVariant === g(ne).Full ? (m(), x(
          "span",
          ot,
          Z(g(t).loading ? g(a)("Input.pauseAnswer") : g(a)("Input.send")),
          1
          /* TEXT */
        )) : _("v-if", !0)
      ]),
      u.value ? (m(), x(
        "div",
        {
          key: 0,
          class: "mc-button-water-wave",
          style: ce(f)
        },
        null,
        4
        /* STYLE */
      )) : _("v-if", !0)
    ], 42, tt));
  }
}), re = /* @__PURE__ */ W(at, [["__scopeId", "data-v-dbbf3b38"]]), st = { class: "mc-input-content" }, lt = {
  key: 0,
  class: "editable-block-container"
}, it = {
  key: 0,
  class: "mc-input-foot"
}, rt = { class: "mc-input-foot-left" }, ut = {
  key: 0,
  class: "mc-input-foot-count"
}, j = 50, ue = /* @__PURE__ */ X({
  __name: "Input",
  props: ke,
  emits: Ee,
  setup(n, { expose: a, emit: s }) {
    const t = n, l = s, i = w(null), r = w(!1), u = w(""), f = ae(() => ({
      "mc-input": !0,
      "mc-input-disabled": t.disabled,
      "mc-input-simple": t.displayType === V.Simple,
      "mc-input-borderless": t.variant === me.BorderLess
    })), h = () => {
      var o;
      r.value ? (o = i.value) == null || o.clearInput() : u.value = "";
    }, C = () => {
      t.autoClear && h();
    }, d = () => {
      var o;
      return r.value ? (o = i.value) == null ? void 0 : o.getInput() : u.value;
    }, T = (o, L, I) => {
      t.disabled || (r.value = !0, setTimeout(() => {
        var M;
        (M = i.value) == null || M.setInputTag(o, L, I);
      }, j));
    }, b = (o) => {
      t.disabled || (r.value = !0, setTimeout(() => {
        var L;
        (L = i.value) == null || L.setText(o);
      }, j));
    }, B = (o) => {
      t.disabled || (r.value = !0, setTimeout(() => {
        var L;
        (L = i.value) == null || L.setMixTags(o);
      }, j));
    }, z = (o, L, I) => {
      t.disabled || (r.value = !0, setTimeout(() => {
        var M;
        (M = i.value) == null || M.openTipTag(o, L, I);
      }, j));
    }, F = () => {
      var o;
      r.value && ((o = i.value) == null || o.closeTipTag());
    }, k = (o) => {
      o || setTimeout(() => {
        r.value = !1;
      }, j), u.value = o, l("change", o);
    }, v = (o) => {
      l("submit", o);
    }, E = (o) => {
      l("blur", o);
    }, H = (o) => {
      l("focus", o);
    };
    return Y(
      () => t.value,
      () => {
        u.value = t.value;
      },
      { immediate: !0 }
    ), Y(
      () => t.formatContentOptions,
      (o) => {
        const L = o == null ? void 0 : o.formatContent, I = L.filter((U) => U.type === "themeTag");
        I.length ? z(I[0].themeTagText, I[0].popoverContent, I[0].clearInput) : F();
        const M = L.filter((U) => U.type !== "themeTag");
        B(M);
      },
      { deep: !0 }
    ), a({ clearInput: h, getInput: d, setInputTag: T, setText: b, setMixTags: B, openTipTag: z, closeTipTag: F }), _e(Q, { inputValue: u, rootProps: t, rootEmits: l, clearInputAfterSubmit: C }), (o, L) => (m(), x(
      "div",
      {
        class: q(f.value)
      },
      [
        A(o.$slots, "head"),
        $("div", st, [
          A(o.$slots, "prefix"),
          _(" 替换Textarea "),
          r.value ? (m(), x("div", lt, [
            _(" 可编辑块 "),
            r.value ? (m(), O(je, {
              key: 0,
              ref_key: "editableBlockRef",
              ref: i,
              placeholder: t.placeholder,
              maxLength: t.maxLength,
              disabled: t.disabled,
              "submit-short-key": t.submitShortKey,
              autofocus: t.autofocus,
              onInput: k,
              onSend: v,
              onBlur: E,
              onFocus: H
            }, {
              themeTag: pe(({ themeTag: I }) => [
                A(o.$slots, "themeTag", { themeTag: I })
              ]),
              _: 3
              /* FORWARDED */
            }, 8, ["placeholder", "maxLength", "disabled", "submit-short-key", "autofocus"])) : _("v-if", !0)
          ])) : _("v-if", !0),
          r.value ? _("v-if", !0) : (m(), O(Re, { key: 1 })),
          A(o.$slots, "suffix"),
          o.displayType === g(V).Simple ? A(o.$slots, "button", { key: 2 }, () => [
            ie(re)
          ]) : _("v-if", !0)
        ]),
        o.displayType === g(V).Full ? (m(), x("div", it, [
          $("div", rt, [
            A(o.$slots, "extra"),
            o.showCount ? (m(), x(
              "span",
              ut,
              Z(u.value.length) + Z(o.maxLength ?? !1 ? `/${o.maxLength}` : ""),
              1
              /* TEXT */
            )) : _("v-if", !0)
          ]),
          A(o.$slots, "button", {}, () => [
            ie(re)
          ])
        ])) : _("v-if", !0)
      ],
      2
      /* CLASS */
    ));
  }
});
ue.install = (n) => {
  n.component("McInput", ue);
};
export {
  ue as McInput
};
