// @matechat/core/FileList/index.js
import "./index.css";
import { defineComponent as c, openBlock as o, createElementBlock as a, normalizeClass as h, createElementVNode as t, createStaticVNode as f, toDisplayString as k, createCommentVNode as C, ref as $, Fragment as x, renderList as J, createBlock as U, resolveDynamicComponent as Q, unref as m, Teleport as X, createVNode as Y, Transition as t2, withCtx as e2, withModifiers as i2 } from "vue";
import { useMcI18n as l2 } from "@matechat/core/Locale";
const r2 = {
  // 文件列表数据
  fileItems: {
    type: Array,
    default: () => []
  },
  // 组件上下文，决定其外观和行为
  context: {
    type: String,
    validator: (i) => ["input", "dialog"].includes(i)
  }
}, s2 = {
  remove: (i) => i && typeof i == "object",
  "retry-upload": (i) => i && typeof i == "object",
  // 下载、预览和重试下载事件
  download: (i, e) => i && typeof i == "object",
  preview: (i) => i && typeof i == "object",
  "retry-download": (i) => i && typeof i == "object"
}, o2 = ["width", "height", "aria-label"], p = /* @__PURE__ */ c({
  __name: "CodeFile",
  props: {
    size: { default: 32 },
    title: {},
    class: {}
  },
  setup(i) {
    const e = i;
    return (d, r) => (o(), a("svg", {
      width: e.size,
      height: e.size,
      viewBox: "0 0 32 32",
      xmlns: "http://www.w3.org/2000/svg",
      "xmlns:xlink": "http://www.w3.org/1999/xlink",
      fill: "none",
      customFrame: "#000000",
      role: "img",
      "aria-label": e.title,
      class: h(e.class)
    }, r[0] || (r[0] = [
      t(
        "mask",
        {
          id: "mask_5",
          width: "24",
          height: "28",
          x: "4",
          y: "2",
          maskUnits: "userSpaceOnUse",
          "mask-type": "outline"
        },
        [
          t("g", { filter: "url(#pixso_custom_mask_type_outline)" }, [
            t("use", { "xlink:href": "#use_5" })
          ])
        ],
        -1
        /* HOISTED */
      ),
      f('<defs><g id="use_5"><path id="蒙版" d="M7 2L20 2L28 10L28 27C28 28.6569 26.6569 30 25 30L7 30C5.34315 30 4 28.6569 4 27L4 5C4 3.34315 5.34315 2 7 2Z" fill="rgb(92,179,0)" fill-rule="evenodd"></path><path id="蒙版" d="M20 2L7 2C5.34315 2 4 3.34315 4 5L4 27C4 28.6569 5.34315 30 7 30L25 30C26.6569 30 28 28.6569 28 27L28 10L20 2ZM19.5858 3L27 10.4142L27 27C27 27.1577 26.9824 27.3113 26.9472 27.4607C26.9217 27.5687 26.8871 27.6745 26.8432 27.7782C26.7974 27.8867 26.7429 27.9896 26.6798 28.0869C26.6037 28.2041 26.5152 28.3132 26.4142 28.4142C26.3132 28.5152 26.2041 28.6037 26.0869 28.6797C25.9896 28.7429 25.8867 28.7974 25.7782 28.8432C25.6745 28.8871 25.5687 28.9217 25.4607 28.9472C25.3113 28.9824 25.1577 29 25 29L7 29C6.84229 29 6.68872 28.9824 6.53931 28.9472C6.43131 28.9217 6.32548 28.8871 6.22182 28.8432C6.11331 28.7973 6.01038 28.7428 5.91305 28.6797C5.79586 28.6037 5.68677 28.5152 5.58579 28.4142C5.48479 28.3132 5.39628 28.2041 5.32027 28.0869C5.25715 27.9896 5.20265 27.8867 5.15676 27.7782C5.11292 27.6745 5.07827 27.5687 5.05282 27.4607C5.01761 27.3113 5 27.1577 5 27L5 5C5 4.84228 5.01761 4.68871 5.05282 4.53929C5.07827 4.4313 5.11292 4.32547 5.15676 4.22182C5.20266 4.1133 5.25717 4.01037 5.32031 3.91303C5.39631 3.79585 5.4848 3.68677 5.58579 3.58579C5.68677 3.4848 5.79585 3.39631 5.91303 3.32031C6.01037 3.25717 6.1133 3.20266 6.22182 3.15676C6.32547 3.11292 6.4313 3.07827 6.53929 3.05282C6.68871 3.01761 6.84228 3 7 3L19.5858 3Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path></g><filter id="pixso_custom_mask_type_outline"><feColorMatrix type="matrix" values="0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 255 0 "></feColorMatrix></filter></defs><g id="代码类文件"><g id="图标/图片文件"><rect id="矩形" width="32" height="32" x="0" y="0" rx="8" fill="rgb(238,240,245)" fill-opacity="0"></rect><rect id="矩形" width="31" height="31" x="0.5" y="0.5" rx="8" stroke="rgb(151,151,151)" stroke-opacity="0" stroke-width="1"></rect><g id="编组 2"><g id="路径" mask="url(#mask_5)"><use xlink:href="#use_5"></use><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L28 10Z" fill="rgb(185,230,131)" fill-rule="evenodd"></path><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L21 3L21 8.5C21 8.56903 21.0122 8.63301 21.0366 8.69194C21.061 8.75087 21.0976 8.80474 21.1464 8.85355C21.1953 8.90237 21.2491 8.93898 21.3081 8.96339C21.367 8.9878 21.431 9 21.5 9L27 9L28 10Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path><g id="ic_public_code"><path id="ic_public_code_2_5" d="M0 0L7.06296 0" stroke="rgb(255,255,255)" stroke-linecap="round" stroke-width="1.5" transform="matrix(0.351112,-0.936333,0.936333,0.351112,14.7598,21.8066)"></path><path id="矢量 772" d="M12 15L9.05221 18.4391C9.0217 18.4747 9.02001 18.5267 9.04813 18.5642L12 22.5" stroke="rgb(255,255,255)" stroke-width="1.5"></path><path id="矢量 772" d="M3 0L0.0522046 3.43909C0.0217021 3.47468 0.0200084 3.52668 0.0481303 3.56417L3 7.5" stroke="rgb(255,255,255)" stroke-width="1.5" transform="matrix(-1,0,0,1,23,15)"></path></g></g></g></g></g>', 2)
    ]), 10, o2));
  }
}), a2 = ["width", "height", "aria-label"], v = /* @__PURE__ */ c({
  __name: "CompressedFile",
  props: {
    size: { default: 32 },
    title: {},
    class: {}
  },
  setup(i) {
    const e = i;
    return (d, r) => (o(), a("svg", {
      width: e.size,
      height: e.size,
      viewBox: "0 0 32 32",
      xmlns: "http://www.w3.org/2000/svg",
      "xmlns:xlink": "http://www.w3.org/1999/xlink",
      fill: "none",
      customFrame: "#000000",
      role: "img",
      "aria-label": e.title,
      class: h(e.class)
    }, r[0] || (r[0] = [
      t(
        "mask",
        {
          id: "mask_1",
          width: "24",
          height: "28",
          x: "4",
          y: "2",
          maskUnits: "userSpaceOnUse",
          "mask-type": "outline"
        },
        [
          t("g", { filter: "url(#pixso_custom_mask_type_outline)" }, [
            t("use", { "xlink:href": "#use_1" })
          ])
        ],
        -1
        /* HOISTED */
      ),
      f('<defs><g id="use_1"><path id="蒙版" d="M7 2L20 2L28 10L28 27C28 28.6569 26.6569 30 25 30L7 30C5.34315 30 4 28.6569 4 27L4 5C4 3.34315 5.34315 2 7 2Z" fill="rgb(255,136,0)" fill-rule="evenodd"></path><path id="蒙版" d="M20 2L7 2C5.34315 2 4 3.34315 4 5L4 27C4 28.6569 5.34315 30 7 30L25 30C26.6569 30 28 28.6569 28 27L28 10L20 2ZM19.5858 3L27 10.4142L27 27C27 27.1577 26.9824 27.3113 26.9472 27.4607C26.9217 27.5687 26.8871 27.6745 26.8432 27.7782C26.7974 27.8867 26.7429 27.9896 26.6798 28.0869C26.6037 28.2041 26.5152 28.3132 26.4142 28.4142C26.3132 28.5152 26.2041 28.6037 26.0869 28.6797C25.9896 28.7429 25.8867 28.7974 25.7782 28.8432C25.6745 28.8871 25.5687 28.9217 25.4607 28.9472C25.3113 28.9824 25.1577 29 25 29L7 29C6.84229 29 6.68872 28.9824 6.53931 28.9472C6.43131 28.9217 6.32548 28.8871 6.22182 28.8432C6.11331 28.7973 6.01038 28.7428 5.91305 28.6797C5.79586 28.6037 5.68677 28.5152 5.58579 28.4142C5.48479 28.3132 5.39628 28.2041 5.32027 28.0869C5.25715 27.9896 5.20265 27.8867 5.15676 27.7782C5.11292 27.6745 5.07827 27.5687 5.05282 27.4607C5.01761 27.3113 5 27.1577 5 27L5 5C5 4.84228 5.01761 4.68871 5.05282 4.53929C5.07827 4.4313 5.11292 4.32547 5.15676 4.22182C5.20266 4.1133 5.25717 4.01037 5.32031 3.91303C5.39631 3.79585 5.4848 3.68677 5.58579 3.58579C5.68677 3.4848 5.79585 3.39631 5.91303 3.32031C6.01037 3.25717 6.1133 3.20266 6.22182 3.15676C6.32547 3.11292 6.4313 3.07827 6.53929 3.05282C6.68871 3.01761 6.84228 3 7 3L19.5858 3Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path></g><filter id="pixso_custom_mask_type_outline"><feColorMatrix type="matrix" values="0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 255 0 "></feColorMatrix></filter></defs><g id="ppt"><g id="图标/ppt文件"><rect id="矩形" width="32" height="32" x="0" y="0" rx="8" fill="rgb(238,240,245)" fill-opacity="0"></rect><rect id="矩形" width="31" height="31" x="0.5" y="0.5" rx="8" stroke="rgb(151,151,151)" stroke-opacity="0" stroke-width="1"></rect><g id="编组 2"><g id="路径" mask="url(#mask_1)"><use xlink:href="#use_1"></use><g id="编组-109"><path id="Rectangle-16-Copy-2" d="M11.8164 9.26886L14.6559 9.26886C14.844 9.26886 14.9966 9.42141 14.9966 9.60959L14.9966 10.7454C14.9966 10.9336 14.844 11.0861 14.6559 11.0861L11.8164 11.0861L11.8164 9.26886Z" fill="rgb(255,255,255)" fill-rule="evenodd"></path><path id="Rectangle-16-Copy-4" d="M11.8164 12.9036L14.6559 12.9036C14.844 12.9036 14.9966 13.0562 14.9966 13.2444L14.9966 14.7209L14.9966 14.7209L11.8164 14.7209L11.8164 12.9036Z" fill="rgb(255,255,255)" fill-rule="evenodd"></path><path id="Rectangle-16-Copy" d="M0 0L2.83945 0C3.02763 0 3.18019 0.152552 3.18019 0.340734L3.18019 1.47652C3.18019 1.6647 3.02763 1.81725 2.83945 1.81725L0 1.81725L0 0Z" fill="rgb(255,255,255)" fill-rule="evenodd" transform="matrix(-1,0,0,1,13.1802,7.45172)"></path><path id="Rectangle-16-Copy-6" d="M0.340734 0L3.18019 0L3.18019 0L3.18019 1.81725L0.340734 1.81725C0.152552 1.81725 0 1.6647 0 1.47652L0 0.340734C0 0.152552 0.152552 0 0.340734 0Z" fill="rgb(255,255,255)" fill-rule="evenodd" transform="matrix(-1,0,0,1,14.9966,5.63434)"></path><path id="Rectangle-16-Copy-6备份-3" d="M0 0L3.18019 0L3.18019 1.81725L0.340734 1.81725C0.152552 1.81725 0 1.6647 0 1.47652L0 0L0 0Z" fill="rgb(255,255,255)" fill-rule="evenodd" transform="matrix(-1,0,0,1,14.9966,2.00006)"></path><path id="Rectangle-16-Copy-6备份-2" d="M0 0L2.83945 0C3.02763 0 3.18019 0.152552 3.18019 0.340734L3.18019 1.47652C3.18019 1.6647 3.02763 1.81725 2.83945 1.81725L0 1.81725L0 0Z" fill="rgb(255,255,255)" fill-rule="evenodd" transform="matrix(-1,0,0,1,13.1802,3.8172)"></path><path id="Rectangle-16-Copy-3" d="M0 0L2.83945 0C3.02763 0 3.18019 0.152552 3.18019 0.340734L3.18019 1.47652C3.18019 1.6647 3.02763 1.81725 2.83945 1.81725L0 1.81725L0 0Z" fill="rgb(255,255,255)" fill-rule="evenodd" transform="matrix(-1,0,0,1,13.1802,11.0862)"></path><path id="Combined-Shape" d="M15 17.687L15 14.72L10.0026 14.72L10.0026 17.687C10.0026 18.0556 10.3015 18.3545 10.6701 18.3545L14.3324 18.3545C14.7011 18.3545 15 18.0556 15 17.687ZM11.2478 15.6313L13.7465 15.6313L13.7465 16.8807C13.7465 17.0689 13.5939 17.2214 13.4058 17.2214L11.5885 17.2214C11.4003 17.2214 11.2478 17.0689 11.2478 16.8807L11.2478 15.6313Z" fill="rgb(255,255,255)" fill-rule="evenodd"></path></g><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L28 10Z" fill="rgb(252,213,164)" fill-rule="evenodd"></path><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L21 3L21 8.5C21 8.56903 21.0122 8.63301 21.0366 8.69194C21.061 8.75087 21.0976 8.80474 21.1464 8.85355C21.1953 8.90237 21.2491 8.93898 21.3081 8.96339C21.367 8.9878 21.431 9 21.5 9L27 9L28 10Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path></g></g></g></g>', 2)
    ]), 10, a2));
  }
}), d2 = ["width", "height", "aria-label"], w = /* @__PURE__ */ c({
  __name: "Document",
  props: {
    size: { default: 32 },
    title: {},
    class: {}
  },
  setup(i) {
    const e = i;
    return (d, r) => (o(), a("svg", {
      width: e.size,
      height: e.size,
      viewBox: "0 0 32 32",
      xmlns: "http://www.w3.org/2000/svg",
      "xmlns:xlink": "http://www.w3.org/1999/xlink",
      fill: "none",
      customFrame: "#000000",
      role: "img",
      "aria-label": e.title,
      class: h(e.class)
    }, r[0] || (r[0] = [
      t(
        "mask",
        {
          id: "mask_10",
          width: "24",
          height: "28",
          x: "4",
          y: "2",
          maskUnits: "userSpaceOnUse",
          "mask-type": "outline"
        },
        [
          t("g", { filter: "url(#pixso_custom_mask_type_outline)" }, [
            t("use", { "xlink:href": "#use_10" })
          ])
        ],
        -1
        /* HOISTED */
      ),
      f('<defs><g id="use_10"><path id="蒙版" d="M7 2L20 2L28 10L28 27C28 28.6569 26.6569 30 25 30L7 30C5.34315 30 4 28.6569 4 27L4 5C4 3.34315 5.34315 2 7 2Z" fill="rgb(57,121,249)" fill-rule="evenodd"></path><path id="蒙版" d="M20 2L7 2C5.34315 2 4 3.34315 4 5L4 27C4 28.6569 5.34315 30 7 30L25 30C26.6569 30 28 28.6569 28 27L28 10L20 2ZM19.5858 3L27 10.4142L27 27C27 27.1577 26.9824 27.3113 26.9472 27.4607C26.9217 27.5687 26.8871 27.6745 26.8432 27.7782C26.7974 27.8867 26.7429 27.9896 26.6798 28.0869C26.6037 28.2041 26.5152 28.3132 26.4142 28.4142C26.3132 28.5152 26.2041 28.6037 26.0869 28.6797C25.9896 28.7429 25.8867 28.7974 25.7782 28.8432C25.6745 28.8871 25.5687 28.9217 25.4607 28.9472C25.3113 28.9824 25.1577 29 25 29L7 29C6.84229 29 6.68872 28.9824 6.53931 28.9472C6.43131 28.9217 6.32548 28.8871 6.22182 28.8432C6.11331 28.7973 6.01038 28.7428 5.91305 28.6797C5.79586 28.6037 5.68677 28.5152 5.58579 28.4142C5.48479 28.3132 5.39628 28.2041 5.32027 28.0869C5.25715 27.9896 5.20265 27.8867 5.15676 27.7782C5.11292 27.6745 5.07827 27.5687 5.05282 27.4607C5.01761 27.3113 5 27.1577 5 27L5 5C5 4.84228 5.01761 4.68871 5.05282 4.53929C5.07827 4.4313 5.11292 4.32547 5.15676 4.22182C5.20266 4.1133 5.25717 4.01037 5.32031 3.91303C5.39631 3.79585 5.4848 3.68677 5.58579 3.58579C5.68677 3.4848 5.79585 3.39631 5.91303 3.32031C6.01037 3.25717 6.1133 3.20266 6.22182 3.15676C6.32547 3.11292 6.4313 3.07827 6.53929 3.05282C6.68871 3.01761 6.84228 3 7 3L19.5858 3Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path></g><filter id="pixso_custom_mask_type_outline"><feColorMatrix type="matrix" values="0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 255 0 "></feColorMatrix></filter></defs><g id="文件"><rect id="矩形" width="32" height="32" x="0" y="0" rx="6" fill="rgb(238,240,245)" fill-opacity="0"></rect><rect id="矩形" width="31" height="31" x="0.5" y="0.5" rx="6" stroke="rgb(151,151,151)" stroke-opacity="0" stroke-width="1"></rect><g id="图标/空文件夹 copy 2" mask="url(#mask_10)"><use xlink:href="#use_10"></use><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L28 10Z" fill="rgb(157,189,252)" fill-rule="evenodd"></path><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L21 3L21 8.5C21 8.56903 21.0122 8.63301 21.0366 8.69194C21.061 8.75087 21.0976 8.80474 21.1464 8.85355C21.1953 8.90237 21.2491 8.93898 21.3081 8.96339C21.367 8.9878 21.431 9 21.5 9L27 9L28 10Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path><g id="编组 29"><circle id="椭圆形" cx="9.25" cy="13" r="0.75" fill="rgb(255,255,255)"></circle><circle id="椭圆形" cx="9.25" cy="13" r="0.25" stroke="rgb(151,151,151)" stroke-opacity="0" stroke-width="1"></circle><circle id="椭圆形" cx="9.25" cy="18" r="0.75" fill="rgb(255,255,255)"></circle><circle id="椭圆形" cx="9.25" cy="18" r="0.25" stroke="rgb(151,151,151)" stroke-opacity="0" stroke-width="1"></circle><circle id="椭圆形" cx="9.25" cy="23" r="0.75" fill="rgb(255,255,255)"></circle><circle id="椭圆形" cx="9.25" cy="23" r="0.25" stroke="rgb(151,151,151)" stroke-opacity="0" stroke-width="1"></circle></g><path id="矢量 777" d="M12 13L23 13" stroke="rgb(255,255,255)" stroke-linecap="round" stroke-width="1.5"></path><path id="矢量 781" d="M12 18L23 18" stroke="rgb(255,255,255)" stroke-linecap="round" stroke-width="1.5"></path><path id="矢量 782" d="M12 23L23 23" stroke="rgb(255,255,255)" stroke-linecap="round" stroke-width="1.5"></path></g></g>', 2)
    ]), 10, d2));
  }
}), n2 = ["width", "height", "aria-label"], Z = /* @__PURE__ */ c({
  __name: "DrawingBoard",
  props: {
    size: { default: 32 },
    title: {},
    class: {}
  },
  setup(i) {
    const e = i;
    return (d, r) => (o(), a("svg", {
      width: e.size,
      height: e.size,
      viewBox: "0 0 32 32",
      xmlns: "http://www.w3.org/2000/svg",
      "xmlns:xlink": "http://www.w3.org/1999/xlink",
      fill: "none",
      customFrame: "#000000",
      role: "img",
      "aria-label": e.title,
      class: h(e.class)
    }, r[0] || (r[0] = [
      t(
        "mask",
        {
          id: "mask_4",
          width: "24",
          height: "28",
          x: "4",
          y: "2",
          maskUnits: "userSpaceOnUse",
          "mask-type": "outline"
        },
        [
          t("g", { filter: "url(#pixso_custom_mask_type_outline)" }, [
            t("use", { "xlink:href": "#use_4" })
          ])
        ],
        -1
        /* HOISTED */
      ),
      f('<defs><g id="use_4"><path id="蒙版" d="M7 2L20 2L28 10L28 27C28 28.6569 26.6569 30 25 30L7 30C5.34315 30 4 28.6569 4 27L4 5C4 3.34315 5.34315 2 7 2Z" fill="rgb(92,179,0)" fill-rule="evenodd"></path><path id="蒙版" d="M20 2L7 2C5.34315 2 4 3.34315 4 5L4 27C4 28.6569 5.34315 30 7 30L25 30C26.6569 30 28 28.6569 28 27L28 10L20 2ZM19.5858 3L27 10.4142L27 27C27 27.1577 26.9824 27.3113 26.9472 27.4607C26.9217 27.5687 26.8871 27.6745 26.8432 27.7782C26.7974 27.8867 26.7429 27.9896 26.6798 28.0869C26.6037 28.2041 26.5152 28.3132 26.4142 28.4142C26.3132 28.5152 26.2041 28.6037 26.0869 28.6797C25.9896 28.7429 25.8867 28.7974 25.7782 28.8432C25.6745 28.8871 25.5687 28.9217 25.4607 28.9472C25.3113 28.9824 25.1577 29 25 29L7 29C6.84229 29 6.68872 28.9824 6.53931 28.9472C6.43131 28.9217 6.32548 28.8871 6.22182 28.8432C6.11331 28.7973 6.01038 28.7428 5.91305 28.6797C5.79586 28.6037 5.68677 28.5152 5.58579 28.4142C5.48479 28.3132 5.39628 28.2041 5.32027 28.0869C5.25715 27.9896 5.20265 27.8867 5.15676 27.7782C5.11292 27.6745 5.07827 27.5687 5.05282 27.4607C5.01761 27.3113 5 27.1577 5 27L5 5C5 4.84228 5.01761 4.68871 5.05282 4.53929C5.07827 4.4313 5.11292 4.32547 5.15676 4.22182C5.20266 4.1133 5.25717 4.01037 5.32031 3.91303C5.39631 3.79585 5.4848 3.68677 5.58579 3.58579C5.68677 3.4848 5.79585 3.39631 5.91303 3.32031C6.01037 3.25717 6.1133 3.20266 6.22182 3.15676C6.32547 3.11292 6.4313 3.07827 6.53929 3.05282C6.68871 3.01761 6.84228 3 7 3L19.5858 3Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path></g><filter id="pixso_custom_mask_type_outline"><feColorMatrix type="matrix" values="0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 255 0 "></feColorMatrix></filter><clipPath id="clipPath_2"><rect width="15" height="16" x="9" y="11" fill="rgb(255,255,255)"></rect></clipPath></defs>', 1),
      t(
        "g",
        { id: "画板" },
        [
          t("rect", {
            id: "矩形",
            width: "32",
            height: "32",
            x: "0",
            y: "0",
            fill: "rgb(0,0,0)",
            "fill-opacity": "0"
          }),
          t("rect", {
            id: "矩形",
            width: "31",
            height: "31",
            x: "0.5",
            y: "0.5",
            stroke: "rgb(151,151,151)",
            "stroke-opacity": "0",
            "stroke-width": "1"
          }),
          t("g", {
            id: "路径",
            mask: "url(#mask_4)"
          }, [
            t("use", { "xlink:href": "#use_4" }),
            t("g", {
              id: "svg",
              "clip-path": "url(#clipPath_2)",
              customFrame: "url(#clipPath_2)"
            }, [
              t("rect", {
                id: "svg",
                width: "15",
                height: "16",
                x: "9",
                y: "11"
              }),
              t("g", { id: "ic_public_office_brush-画笔/base/ic_public_office_brush1" }, [
                t("path", {
                  id: "path6",
                  d: "M13.5333 23.4382L9.875 24.8333L11.4169 20.7952L19.1124 13.1001C19.8016 12.4109 20.919 12.411 21.6081 13.1002C22.2973 13.7894 22.2973 14.9069 21.6081 15.5961L14.8912 22.3175",
                  stroke: "rgb(255,255,255)",
                  "stroke-linecap": "round",
                  "stroke-linejoin": "round",
                  "stroke-width": "1.5"
                }),
                t("line", {
                  id: "path7",
                  x1: "0",
                  x2: "3.51560974",
                  y1: "0",
                  y2: "0",
                  stroke: "rgb(255,255,255)",
                  "stroke-linecap": "square",
                  "stroke-linejoin": "round",
                  "stroke-width": "1.5",
                  transform: "matrix(0.705038,0.70917,-0.70917,0.705038,18.5347,13.6929)"
                }),
                t("path", {
                  id: "path8",
                  d: "M16.3501 23.632C16.7621 24.4423 17.2822 24.8475 17.9105 24.8475C18.8918 24.8475 19.3342 23.6954 20.1977 23.632C20.7734 23.5898 21.3469 23.8969 21.9183 24.5533",
                  stroke: "rgb(255,255,255)",
                  "stroke-linecap": "round",
                  "stroke-linejoin": "round",
                  "stroke-width": "1.5"
                }),
                t("path", {
                  id: "矢量 793",
                  d: "M14 25.0878C14.1434 25.1785 14.5505 25.3054 15.0323 25.0878C15.514 24.8703 17.7849 24 18.2151 24C18.6452 24 18.7828 24.272 18.6452 24.6346C18.5075 24.9972 17.957 25.6318 18.129 25.9037C18.3011 26.1756 19.9355 25.8131 20.6237 25.5411C21.3118 25.2691 21.828 25.3598 22 25.5411",
                  opacity: "0",
                  stroke: "rgb(255,255,255)",
                  "stroke-width": "1.5"
                }),
                t("path", {
                  id: "path9",
                  d: "M20.1849 23.4575C20.8232 23.4106 21.4459 23.7441 22.0503 24.4384C22.1137 24.5113 22.1061 24.6219 22.0332 24.6853C21.9603 24.7488 21.8497 24.7411 21.7863 24.6682C21.2478 24.0496 20.7235 23.7689 20.2105 23.8065C19.9376 23.8265 19.7196 23.9568 19.338 24.287C19.2903 24.3284 19.2903 24.3284 19.2421 24.3704C18.6985 24.8433 18.3888 25.0225 17.9105 25.0225C17.2076 25.0225 16.6332 24.5751 16.1941 23.7113C16.1503 23.6251 16.1846 23.5198 16.2708 23.476C16.3569 23.4322 16.4623 23.4665 16.5061 23.5527C16.8908 24.3095 17.3568 24.6725 17.9105 24.6725C18.2771 24.6725 18.5295 24.5264 19.0123 24.1063C19.0605 24.0644 19.0605 24.0644 19.109 24.0223C19.5446 23.6454 19.8136 23.4847 20.1849 23.4575Z",
                  fill: "rgb(255,255,255)",
                  "fill-opacity": "0",
                  "fill-rule": "nonzero"
                }),
                t("path", {
                  id: "path9",
                  d: "M22.0503 24.4384C22.1137 24.5113 22.1061 24.6219 22.0332 24.6853C21.9603 24.7488 21.8497 24.7411 21.7863 24.6682C21.2478 24.0496 20.7235 23.7689 20.2105 23.8065C19.9376 23.8265 19.7196 23.9568 19.338 24.287C19.2903 24.3284 19.2903 24.3284 19.2421 24.3704C18.6985 24.8433 18.3888 25.0225 17.9105 25.0225C17.2076 25.0225 16.6332 24.5751 16.1941 23.7113C16.1503 23.6251 16.1846 23.5198 16.2708 23.476C16.3569 23.4322 16.4623 23.4665 16.5061 23.5527C16.8908 24.3095 17.3568 24.6725 17.9105 24.6725C18.2771 24.6725 18.5295 24.5264 19.0123 24.1063C19.0605 24.0644 19.0605 24.0644 19.109 24.0223C19.5446 23.6454 19.8136 23.4847 20.1849 23.4575C20.8232 23.4106 21.4459 23.7441 22.0503 24.4384Z",
                  "fill-rule": "nonzero",
                  stroke: "rgb(255,255,255)",
                  "stroke-opacity": "0",
                  "stroke-linejoin": "round",
                  "stroke-width": "0.699999988"
                })
              ])
            ]),
            t("path", {
              id: "路径",
              d: "M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L28 10Z",
              fill: "rgb(185,230,131)",
              "fill-rule": "evenodd"
            }),
            t("path", {
              id: "路径",
              d: "M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L21 3L21 8.5C21 8.56903 21.0122 8.63301 21.0366 8.69194C21.061 8.75087 21.0976 8.80474 21.1464 8.85355C21.1953 8.90237 21.2491 8.93898 21.3081 8.96339C21.367 8.9878 21.431 9 21.5 9L27 9L28 10Z",
              fill: "rgb(223,225,230)",
              "fill-opacity": "0",
              "fill-rule": "evenodd"
            }),
            t("path", {
              id: "路径 6",
              d: "M8.5 21.3938L12.2173 17.4814L16.3341 22L23.5 14L8.5 21.3938Z",
              opacity: "0",
              fill: "rgb(216,216,216)",
              "fill-opacity": "0",
              "fill-rule": "evenodd"
            }),
            t("path", {
              id: "路径 6",
              d: "M8.5 21.3938L12.2173 17.4814L16.3341 22L23.5 14",
              opacity: "0",
              "fill-rule": "evenodd",
              stroke: "rgb(255,255,255)",
              "stroke-width": "1.5"
            })
          ])
        ],
        -1
        /* HOISTED */
      )
    ]), 10, n2));
  }
}), p2 = ["width", "height", "aria-label"], C2 = /* @__PURE__ */ c({
  __name: "EmailFile",
  props: {
    size: { default: 32 },
    title: {},
    class: {}
  },
  setup(i) {
    const e = i;
    return (d, r) => (o(), a("svg", {
      width: e.size,
      height: e.size,
      viewBox: "0 0 32 32",
      xmlns: "http://www.w3.org/2000/svg",
      "xmlns:xlink": "http://www.w3.org/1999/xlink",
      fill: "none",
      customFrame: "#000000",
      role: "img",
      "aria-label": e.title,
      class: h(e.class)
    }, r[0] || (r[0] = [
      t(
        "mask",
        {
          id: "mask_14",
          width: "28",
          height: "22",
          x: "2",
          y: "5",
          maskUnits: "userSpaceOnUse",
          "mask-type": "outline"
        },
        [
          t("g", { filter: "url(#pixso_custom_mask_type_outline)" }, [
            t("use", { "xlink:href": "#use_14" })
          ])
        ],
        -1
        /* HOISTED */
      ),
      f('<defs><g id="use_14"><rect id="蒙版" width="28" height="22" x="2" y="5" rx="4" fill="rgb(255,136,0)"></rect><rect id="蒙版" width="27" height="21" x="2.5" y="5.5" rx="4" stroke="rgb(223,225,230)" stroke-opacity="0" stroke-width="1"></rect></g><filter id="pixso_custom_mask_type_outline"><feColorMatrix type="matrix" values="0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 255 0 "></feColorMatrix></filter><linearGradient id="paint_linear_0" x1="14.3386717" x2="14.3386717" y1="0" y2="11.6952209" gradientUnits="userSpaceOnUse"><stop stop-color="rgb(246,247,253)" offset="0" stop-opacity="1"></stop><stop stop-color="rgb(233,237,250)" offset="1" stop-opacity="1"></stop></linearGradient><linearGradient id="paint_linear_1" x1="16" x2="16" y1="3" y2="17" gradientUnits="userSpaceOnUse"><stop stop-color="rgb(246,247,253)" offset="0" stop-opacity="1"></stop><stop stop-color="rgb(233,237,250)" offset="1" stop-opacity="1"></stop></linearGradient></defs><g id="邮件"><rect id="矩形" width="32" height="32" x="0" y="0" rx="4" fill="rgb(238,240,245)" fill-opacity="0"></rect><rect id="矩形" width="31" height="31" x="0.5" y="0.5" rx="4" stroke="rgb(151,151,151)" stroke-opacity="0" stroke-width="1"></rect><g id="矩形" mask="url(#mask_14)"><use xlink:href="#use_14"></use><path id="形状" d="M0.341524 1.75256L11.7046 11.6952L0.999972 0C0.711642 0 0.43735 0.124481 0.247408 0.341492C-0.116239 0.757141 -0.074125 1.38889 0.341524 1.75256ZM27.6773 0C28.2296 0 28.6773 0.447723 28.6773 1C28.6773 1.28836 28.5528 1.56268 28.3358 1.75256L16.9727 11.6952L27.6773 0Z" opacity="0" fill="url(#paint_linear_0)" fill-opacity="0" fill-rule="evenodd" transform="matrix(1,0,0,-1,1.66125,27.4999)"></path><path id="形状" d="M0.341524 1.75256L11.7046 11.6952L0.999972 0C0.711642 0 0.43735 0.124481 0.247408 0.341492C-0.116239 0.757141 -0.074125 1.38889 0.341524 1.75256ZM27.6773 0C28.2296 0 28.6773 0.447723 28.6773 1C28.6773 1.28836 28.5528 1.56268 28.3358 1.75256L16.9727 11.6952L27.6773 0Z" opacity="0" fill-rule="evenodd" stroke="rgb(255,255,255)" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" transform="matrix(1,0,0,-1,1.66125,27.4999)"></path><path id="矩形" d="M2.66145 3L29.3386 3C30.2619 3 30.692 4.14454 29.9971 4.75258L18.634 14.6952C17.1259 16.0148 14.8741 16.0148 13.366 14.6952L2.00294 4.75258C1.30804 4.14454 1.73808 3 2.66145 3Z" fill="url(#paint_linear_1)" fill-opacity="0" fill-rule="evenodd"></path><path id="矩形" d="M29.3386 3C30.2619 3 30.692 4.14454 29.9971 4.75258L18.634 14.6952C17.1259 16.0148 14.8741 16.0148 13.366 14.6952L2.00294 4.75258C1.30804 4.14454 1.73808 3 2.66145 3L29.3386 3Z" fill-rule="evenodd" stroke="rgb(255,255,255)" stroke-width="1.5"></path><path id="矢量 789" d="M0 0L15.5563 0" stroke="rgb(255,255,255)" stroke-width="1.5" transform="matrix(0.707107,-0.707107,0.707107,0.707107,3,26)"></path><path id="矢量 789" d="M0 0L15.5563 0" stroke="rgb(255,255,255)" stroke-width="1.5" transform="matrix(-0.707107,-0.707107,-0.707107,0.707107,29,26)"></path></g></g>', 2)
    ]), 10, p2));
  }
}), h2 = ["width", "height", "aria-label"], B = /* @__PURE__ */ c({
  __name: "Excel",
  props: {
    size: { default: 32 },
    title: {},
    class: {}
  },
  setup(i) {
    const e = i;
    return (d, r) => (o(), a("svg", {
      width: e.size,
      height: e.size,
      viewBox: "0 0 32 32",
      xmlns: "http://www.w3.org/2000/svg",
      "xmlns:xlink": "http://www.w3.org/1999/xlink",
      fill: "none",
      customFrame: "#000000",
      role: "img",
      "aria-label": e.title,
      class: h(e.class)
    }, r[0] || (r[0] = [
      t(
        "mask",
        {
          id: "mask_6",
          width: "24",
          height: "28",
          x: "4",
          y: "2",
          maskUnits: "userSpaceOnUse",
          "mask-type": "outline"
        },
        [
          t("g", { filter: "url(#pixso_custom_mask_type_outline)" }, [
            t("use", { "xlink:href": "#use_6" })
          ])
        ],
        -1
        /* HOISTED */
      ),
      f('<defs><g id="use_6"><path id="蒙版" d="M7 2L20 2L28 10L28 27C28 28.6569 26.6569 30 25 30L7 30C5.34315 30 4 28.6569 4 27L4 5C4 3.34315 5.34315 2 7 2Z" fill="rgb(37,194,81)" fill-rule="evenodd"></path><path id="蒙版" d="M20 2L7 2C5.34315 2 4 3.34315 4 5L4 27C4 28.6569 5.34315 30 7 30L25 30C26.6569 30 28 28.6569 28 27L28 10L20 2ZM19.5858 3L27 10.4142L27 27C27 27.1577 26.9824 27.3113 26.9472 27.4607C26.9217 27.5687 26.8871 27.6745 26.8432 27.7782C26.7974 27.8867 26.7429 27.9896 26.6798 28.0869C26.6037 28.2041 26.5152 28.3132 26.4142 28.4142C26.3132 28.5152 26.2041 28.6037 26.0869 28.6797C25.9896 28.7429 25.8867 28.7974 25.7782 28.8432C25.6745 28.8871 25.5687 28.9217 25.4607 28.9472C25.3113 28.9824 25.1577 29 25 29L7 29C6.84229 29 6.68872 28.9824 6.53931 28.9472C6.43131 28.9217 6.32548 28.8871 6.22182 28.8432C6.11331 28.7973 6.01038 28.7428 5.91305 28.6797C5.79586 28.6037 5.68677 28.5152 5.58579 28.4142C5.48479 28.3132 5.39628 28.2041 5.32027 28.0869C5.25715 27.9896 5.20265 27.8867 5.15676 27.7782C5.11292 27.6745 5.07827 27.5687 5.05282 27.4607C5.01761 27.3113 5 27.1577 5 27L5 5C5 4.84228 5.01761 4.68871 5.05282 4.53929C5.07827 4.4313 5.11292 4.32547 5.15676 4.22182C5.20266 4.1133 5.25717 4.01037 5.32031 3.91303C5.39631 3.79585 5.4848 3.68677 5.58579 3.58579C5.68677 3.4848 5.79585 3.39631 5.91303 3.32031C6.01037 3.25717 6.1133 3.20266 6.22182 3.15676C6.32547 3.11292 6.4313 3.07827 6.53929 3.05282C6.68871 3.01761 6.84228 3 7 3L19.5858 3Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path></g><filter id="pixso_custom_mask_type_outline"><feColorMatrix type="matrix" values="0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 255 0 "></feColorMatrix></filter></defs><g id="表格"><rect id="矩形" width="32" height="32" x="0" y="0" fill="rgb(0,0,0)" fill-opacity="0"></rect><rect id="矩形" width="31" height="31" x="0.5" y="0.5" stroke="rgb(151,151,151)" stroke-opacity="0" stroke-width="1"></rect><g id="路径" mask="url(#mask_6)"><use xlink:href="#use_6"></use><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L28 10Z" fill="rgb(146,224,165)" fill-rule="evenodd"></path><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L21 3L21 8.5C21 8.56903 21.0122 8.63301 21.0366 8.69194C21.061 8.75087 21.0976 8.80474 21.1464 8.85355C21.1953 8.90237 21.2491 8.93898 21.3081 8.96339C21.367 8.9878 21.431 9 21.5 9L27 9L28 10Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path><g id="组合 3851"><rect id="矩形" width="15" height="13" x="8.5" y="12.5" rx="2" fill="rgb(216,216,216)" fill-opacity="0"></rect><rect id="矩形" width="15" height="13" x="8.5" y="12.5" rx="2" stroke="rgb(255,255,255)" stroke-width="1.5"></rect><path id="矢量 773" d="M9 19L23 19" stroke="rgb(255,255,255)" stroke-width="1.5"></path><path id="矢量 773" d="M0 0L14 0" stroke="rgb(255,255,255)" stroke-width="1.5" transform="matrix(0,-1,1,0,16,26)"></path></g></g></g>', 2)
    ]), 10, h2));
  }
}), c2 = ["width", "height", "aria-label"], u2 = /* @__PURE__ */ c({
  __name: "FlowChart",
  props: {
    size: { default: 32 },
    title: {},
    class: {}
  },
  setup(i) {
    const e = i;
    return (d, r) => (o(), a("svg", {
      width: e.size,
      height: e.size,
      viewBox: "0 0 32 32",
      xmlns: "http://www.w3.org/2000/svg",
      "xmlns:xlink": "http://www.w3.org/1999/xlink",
      fill: "none",
      customFrame: "#000000",
      role: "img",
      "aria-label": e.title,
      class: h(e.class)
    }, r[0] || (r[0] = [
      t(
        "mask",
        {
          id: "mask_2",
          width: "24",
          height: "28",
          x: "4",
          y: "2",
          maskUnits: "userSpaceOnUse",
          "mask-type": "outline"
        },
        [
          t("g", { filter: "url(#pixso_custom_mask_type_outline)" }, [
            t("use", { "xlink:href": "#use_2" })
          ])
        ],
        -1
        /* HOISTED */
      ),
      f('<defs><g id="use_2"><path id="蒙版" d="M7 2L20 2L28 10L28 27C28 28.6569 26.6569 30 25 30L7 30C5.34315 30 4 28.6569 4 27L4 5C4 3.34315 5.34315 2 7 2Z" fill="rgb(242,48,48)" fill-rule="evenodd"></path><path id="蒙版" d="M20 2L7 2C5.34315 2 4 3.34315 4 5L4 27C4 28.6569 5.34315 30 7 30L25 30C26.6569 30 28 28.6569 28 27L28 10L20 2ZM19.5858 3L27 10.4142L27 27C27 27.1577 26.9824 27.3113 26.9472 27.4607C26.9217 27.5687 26.8871 27.6745 26.8432 27.7782C26.7974 27.8867 26.7429 27.9896 26.6798 28.0869C26.6037 28.2041 26.5152 28.3132 26.4142 28.4142C26.3132 28.5152 26.2041 28.6037 26.0869 28.6797C25.9896 28.7429 25.8867 28.7974 25.7782 28.8432C25.6745 28.8871 25.5687 28.9217 25.4607 28.9472C25.3113 28.9824 25.1577 29 25 29L7 29C6.84229 29 6.68872 28.9824 6.53931 28.9472C6.43131 28.9217 6.32548 28.8871 6.22182 28.8432C6.11331 28.7973 6.01038 28.7428 5.91305 28.6797C5.79586 28.6037 5.68677 28.5152 5.58579 28.4142C5.48479 28.3132 5.39628 28.2041 5.32027 28.0869C5.25715 27.9896 5.20265 27.8867 5.15676 27.7782C5.11292 27.6745 5.07827 27.5687 5.05282 27.4607C5.01761 27.3113 5 27.1577 5 27L5 5C5 4.84228 5.01761 4.68871 5.05282 4.53929C5.07827 4.4313 5.11292 4.32547 5.15676 4.22182C5.20266 4.1133 5.25717 4.01037 5.32031 3.91303C5.39631 3.79585 5.4848 3.68677 5.58579 3.58579C5.68677 3.4848 5.79585 3.39631 5.91303 3.32031C6.01037 3.25717 6.1133 3.20266 6.22182 3.15676C6.32547 3.11292 6.4313 3.07827 6.53929 3.05282C6.68871 3.01761 6.84228 3 7 3L19.5858 3Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path></g><filter id="pixso_custom_mask_type_outline"><feColorMatrix type="matrix" values="0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 255 0 "></feColorMatrix></filter></defs><g id="流程图"><rect id="矩形" width="32" height="32" x="0" y="0" fill="rgb(0,0,0)" fill-opacity="0"></rect><rect id="矩形" width="31" height="31" x="0.5" y="0.5" stroke="rgb(151,151,151)" stroke-opacity="0" stroke-width="1"></rect><g id="编组 2" mask="url(#mask_2)"><use xlink:href="#use_2"></use><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L28 10Z" fill="rgb(250,167,163)" fill-rule="evenodd"></path><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L21 3L21 8.5C21 8.56903 21.0122 8.63301 21.0366 8.69194C21.061 8.75087 21.0976 8.80474 21.1464 8.85355C21.1953 8.90237 21.2491 8.93898 21.3081 8.96339C21.367 8.9878 21.431 9 21.5 9L27 9L28 10Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path><path id="路径" d="M6.54991 5L2.5 5C1.11929 5 -1.09279e-07 3.88071 -1.09279e-07 2.5C-1.09279e-07 1.11929 1.11929 0 2.5 0L12 0L6.54991 5Z" fill="rgb(216,216,216)" fill-opacity="0" fill-rule="evenodd" transform="matrix(-1,0,0,1,23.5,13)"></path><path id="路径" d="M6.54991 5L2.5 5C1.11929 5 -1.09279e-07 3.88071 -1.09279e-07 2.5C-1.09279e-07 1.11929 1.11929 0 2.5 0L12 0" fill-rule="evenodd" stroke="rgb(255,255,255)" stroke-width="1.5" transform="matrix(-1,0,0,1,23.5,13)"></path><path id="路径备份" d="M8.64563 5L2.5 5C1.11929 5 -1.09278e-07 3.88071 -1.09278e-07 2.5C-1.09278e-07 1.11929 1.11929 0 2.5 0L11.9084 0L8.64563 5Z" fill="rgb(216,216,216)" fill-opacity="0" fill-rule="evenodd" transform="matrix(1,0,0,-1,9,23)"></path><path id="路径备份" d="M8.64563 5L2.5 5C1.11929 5 -1.09278e-07 3.88071 -1.09278e-07 2.5C-1.09278e-07 1.11929 1.11929 0 2.5 0L11.9084 0" fill-rule="evenodd" stroke="rgb(255,255,255)" stroke-width="1.5" transform="matrix(1,0,0,-1,9,23)"></path><rect id="矩形" width="3" height="3" x="20.5" y="21.5" rx="1.5" fill="rgb(18,119,255)" fill-opacity="0"></rect><rect id="矩形" width="3" height="3" x="20.5" y="21.5" rx="1.5" stroke="rgb(255,255,255)" stroke-width="1.5"></rect><rect id="矩形备份 16" width="3" height="3" x="0" y="0" rx="1.5" fill="rgb(216,216,216)" fill-opacity="0" transform="matrix(0.707107,0.707107,-0.707107,0.707107,10.5,10.8787)"></rect><rect id="矩形备份 16" width="3" height="3" x="0" y="0" rx="1.5" stroke="rgb(255,255,255)" stroke-width="1.5" transform="matrix(0.707107,0.707107,-0.707107,0.707107,10.5,10.8787)"></rect></g></g>', 2)
    ]), 10, c2));
  }
}), f2 = ["width", "height", "aria-label"], y = /* @__PURE__ */ c({
  __name: "Image",
  props: {
    size: { default: 32 },
    title: {},
    class: {}
  },
  setup(i) {
    const e = i;
    return (d, r) => (o(), a("svg", {
      width: e.size,
      height: e.size,
      viewBox: "0 0 32 32",
      xmlns: "http://www.w3.org/2000/svg",
      "xmlns:xlink": "http://www.w3.org/1999/xlink",
      fill: "none",
      customFrame: "#000000",
      role: "img",
      "aria-label": e.title,
      class: h(e.class)
    }, r[0] || (r[0] = [
      t(
        "mask",
        {
          id: "mask_0",
          width: "24",
          height: "28",
          x: "4",
          y: "2",
          maskUnits: "userSpaceOnUse",
          "mask-type": "outline"
        },
        [
          t("g", { filter: "url(#pixso_custom_mask_type_outline)" }, [
            t("use", { "xlink:href": "#use_0" })
          ])
        ],
        -1
        /* HOISTED */
      ),
      f('<defs><g id="use_0"><path id="蒙版" d="M7 2L20 2L28 10L28 27C28 28.6569 26.6569 30 25 30L7 30C5.34315 30 4 28.6569 4 27L4 5C4 3.34315 5.34315 2 7 2Z" fill="rgb(57,121,249)" fill-rule="evenodd"></path><path id="蒙版" d="M20 2L7 2C5.34315 2 4 3.34315 4 5L4 27C4 28.6569 5.34315 30 7 30L25 30C26.6569 30 28 28.6569 28 27L28 10L20 2ZM19.5858 3L27 10.4142L27 27C27 27.1577 26.9824 27.3113 26.9472 27.4607C26.9217 27.5687 26.8871 27.6745 26.8432 27.7782C26.7974 27.8867 26.7429 27.9896 26.6798 28.0869C26.6037 28.2041 26.5152 28.3132 26.4142 28.4142C26.3132 28.5152 26.2041 28.6037 26.0869 28.6797C25.9896 28.7429 25.8867 28.7974 25.7782 28.8432C25.6745 28.8871 25.5687 28.9217 25.4607 28.9472C25.3113 28.9824 25.1577 29 25 29L7 29C6.84229 29 6.68872 28.9824 6.53931 28.9472C6.43131 28.9217 6.32548 28.8871 6.22182 28.8432C6.11331 28.7973 6.01038 28.7428 5.91305 28.6797C5.79586 28.6037 5.68677 28.5152 5.58579 28.4142C5.48479 28.3132 5.39628 28.2041 5.32027 28.0869C5.25715 27.9896 5.20265 27.8867 5.15676 27.7782C5.11292 27.6745 5.07827 27.5687 5.05282 27.4607C5.01761 27.3113 5 27.1577 5 27L5 5C5 4.84228 5.01761 4.68871 5.05282 4.53929C5.07827 4.4313 5.11292 4.32547 5.15676 4.22182C5.20266 4.1133 5.25717 4.01037 5.32031 3.91303C5.39631 3.79585 5.4848 3.68677 5.58579 3.58579C5.68677 3.4848 5.79585 3.39631 5.91303 3.32031C6.01037 3.25717 6.1133 3.20266 6.22182 3.15676C6.32547 3.11292 6.4313 3.07827 6.53929 3.05282C6.68871 3.01761 6.84228 3 7 3L19.5858 3Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path></g><filter id="pixso_custom_mask_type_outline"><feColorMatrix type="matrix" values="0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 255 0 "></feColorMatrix></filter></defs><g id="图片"><rect id="矩形" width="32" height="32" x="0" y="0" rx="6" fill="rgb(238,240,245)" fill-opacity="0"></rect><rect id="矩形" width="31" height="31" x="0.5" y="0.5" rx="6" stroke="rgb(151,151,151)" stroke-opacity="0" stroke-width="1"></rect><g id="图标/空文件夹 copy 2" mask="url(#mask_0)"><use xlink:href="#use_0"></use><g id="编组 27"><path id="椭圆形" d="M11.5 16.5C12.3284 16.5 13 15.8284 13 15C13 14.1716 12.3284 13.5 11.5 13.5C10.6716 13.5 10 14.1716 10 15C10 15.8284 10.6716 16.5 11.5 16.5Z" fill="rgb(255,255,255)" fill-rule="evenodd"></path><path id="椭圆形" d="M13 15C13 15.8284 12.3284 16.5 11.5 16.5C10.6716 16.5 10 15.8284 10 15C10 14.1716 10.6716 13.5 11.5 13.5C12.3284 13.5 13 14.1716 13 15ZM11.9634 14.8081C11.9878 14.867 12 14.931 12 15C12 15.069 11.9878 15.133 11.9634 15.1919C11.939 15.2509 11.9024 15.3047 11.8536 15.3536C11.8047 15.4024 11.7509 15.439 11.6919 15.4634C11.633 15.4878 11.569 15.5 11.5 15.5C11.431 15.5 11.367 15.4878 11.3081 15.4634C11.2491 15.439 11.1953 15.4024 11.1464 15.3536C11.0976 15.3047 11.061 15.2509 11.0366 15.1919C11.0122 15.133 11 15.069 11 15C11 14.931 11.0122 14.867 11.0366 14.8081C11.061 14.7491 11.0976 14.6953 11.1464 14.6464C11.1953 14.5976 11.2491 14.561 11.3081 14.5366C11.367 14.5122 11.431 14.5 11.5 14.5C11.569 14.5 11.633 14.5122 11.6919 14.5366C11.7509 14.561 11.8047 14.5976 11.8536 14.6464C11.9024 14.6953 11.939 14.7491 11.9634 14.8081Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path><path id="路径" d="M22.7002 23.126L18.6447 16.9265C18.3065 16.4096 17.5571 16.3873 17.1888 16.8831L15.2857 19.4454L13.7373 18.0224C13.3287 17.647 12.6798 17.734 12.3846 18.2038L9.28364 23.1397C8.91213 23.731 9.33716 24.5 10.0355 24.5L21.9572 24.5C22.6629 24.5 23.0866 23.7165 22.7002 23.126Z" fill="rgb(255,255,255)" fill-rule="evenodd"></path><path id="路径" d="M14.5454 18.765L13.7373 18.0224C13.3287 17.647 12.6798 17.734 12.3846 18.2038L9.28364 23.1397C8.91213 23.731 9.33716 24.5 10.0355 24.5L21.9572 24.5C22.6629 24.5 23.0866 23.7165 22.7002 23.126L18.6447 16.9265C18.3065 16.4096 17.5571 16.3873 17.1888 16.8831L15.8852 18.6382L15.2857 19.4454L14.5454 18.765ZM17.8958 17.6084L15.4265 20.9329L13.1597 18.8498L10.2383 23.5L21.7499 23.5L17.8958 17.6084Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path></g><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L28 10Z" fill="rgb(157,189,252)" fill-rule="evenodd"></path><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L21 3L21 8.5C21 8.56903 21.0122 8.63301 21.0366 8.69194C21.061 8.75087 21.0976 8.80474 21.1464 8.85355C21.1953 8.90237 21.2491 8.93898 21.3081 8.96339C21.367 8.9878 21.431 9 21.5 9L27 9L28 10Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path></g></g>', 2)
    ]), 10, f2));
  }
}), g2 = ["width", "height", "aria-label"], W = /* @__PURE__ */ c({
  __name: "Markdown",
  props: {
    size: { default: 32 },
    title: {},
    class: {}
  },
  setup(i) {
    const e = i;
    return (d, r) => (o(), a("svg", {
      width: e.size,
      height: e.size,
      viewBox: "0 0 32 32",
      xmlns: "http://www.w3.org/2000/svg",
      "xmlns:xlink": "http://www.w3.org/1999/xlink",
      fill: "none",
      customFrame: "#000000",
      role: "img",
      "aria-label": e.title,
      class: h(e.class)
    }, r[0] || (r[0] = [
      t(
        "mask",
        {
          id: "mask_18",
          width: "24",
          height: "28",
          x: "4",
          y: "2",
          maskUnits: "userSpaceOnUse",
          "mask-type": "outline"
        },
        [
          t("g", { filter: "url(#pixso_custom_mask_type_outline)" }, [
            t("use", { "xlink:href": "#use_18" })
          ])
        ],
        -1
        /* HOISTED */
      ),
      f('<defs><g id="use_18"><path id="蒙版" d="M7 2L20 2L28 10L28 27C28 28.6569 26.6569 30 25 30L7 30C5.34315 30 4 28.6569 4 27L4 5C4 3.34315 5.34315 2 7 2Z" fill="rgb(37,194,81)" fill-rule="evenodd"></path><path id="蒙版" d="M20 2L7 2C5.34315 2 4 3.34315 4 5L4 27C4 28.6569 5.34315 30 7 30L25 30C26.6569 30 28 28.6569 28 27L28 10L20 2ZM19.5858 3L27 10.4142L27 27C27 27.1577 26.9824 27.3113 26.9472 27.4607C26.9217 27.5687 26.8871 27.6745 26.8432 27.7782C26.7974 27.8867 26.7429 27.9896 26.6798 28.0869C26.6037 28.2041 26.5152 28.3132 26.4142 28.4142C26.3132 28.5152 26.2041 28.6037 26.0869 28.6797C25.9896 28.7429 25.8867 28.7974 25.7782 28.8432C25.6745 28.8871 25.5687 28.9217 25.4607 28.9472C25.3113 28.9824 25.1577 29 25 29L7 29C6.84229 29 6.68872 28.9824 6.53931 28.9472C6.43131 28.9217 6.32548 28.8871 6.22182 28.8432C6.11331 28.7973 6.01038 28.7428 5.91305 28.6797C5.79586 28.6037 5.68677 28.5152 5.58579 28.4142C5.48479 28.3132 5.39628 28.2041 5.32027 28.0869C5.25715 27.9896 5.20265 27.8867 5.15676 27.7782C5.11292 27.6745 5.07827 27.5687 5.05282 27.4607C5.01761 27.3113 5 27.1577 5 27L5 5C5 4.84228 5.01761 4.68871 5.05282 4.53929C5.07827 4.4313 5.11292 4.32547 5.15676 4.22182C5.20266 4.1133 5.25717 4.01037 5.32031 3.91303C5.39631 3.79585 5.4848 3.68677 5.58579 3.58579C5.68677 3.4848 5.79585 3.39631 5.91303 3.32031C6.01037 3.25717 6.1133 3.20266 6.22182 3.15676C6.32547 3.11292 6.4313 3.07827 6.53929 3.05282C6.68871 3.01761 6.84228 3 7 3L19.5858 3Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path></g><filter id="pixso_custom_mask_type_outline"><feColorMatrix type="matrix" values="0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 255 0 "></feColorMatrix></filter></defs><rect id="雪碧图/32px/markdown" width="32" height="32" x="0" y="0"></rect><g id="图标/markdown文件"><rect id="矩形" width="32" height="32" x="0" y="0" rx="4" fill="rgb(238,240,245)" fill-opacity="0"></rect><rect id="矩形" width="31" height="31" x="0.5" y="0.5" rx="4" stroke="rgb(151,151,151)" stroke-opacity="0" stroke-width="1"></rect><g id="编组 2" mask="url(#mask_18)"><use xlink:href="#use_18"></use><g id="组合 3854"><path id="矢量 770" d="M0 0L8 0" stroke="rgb(255,255,255)" stroke-linecap="round" stroke-width="1.5" transform="matrix(0,1,-1,0,9,14)"></path><path id="矢量 790" d="M0 0L8 0" stroke="rgb(255,255,255)" stroke-linecap="round" stroke-width="1.5" transform="matrix(0,1,-1,0,17,14)"></path><path id="矢量 794" d="M0 0L8 0" stroke="rgb(255,255,255)" stroke-linecap="round" stroke-width="1.5" transform="matrix(0,1,-1,0,22,14)"></path><path id="矢量 795" d="M20 20L21.9293 21.9293C21.9683 21.9683 22.0317 21.9683 22.0707 21.9293L24 20" stroke="rgb(255,255,255)" stroke-width="1.5"></path><path id="矢量 791" d="M9 14L12.9106 21.8211C12.9474 21.8948 13.0526 21.8948 13.0894 21.8211L17 14" stroke="rgb(255,255,255)" stroke-width="1.5"></path></g><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L28 10Z" fill="rgb(146,224,165)" fill-rule="evenodd"></path><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L21 3L21 8.5C21 8.56903 21.0122 8.63301 21.0366 8.69194C21.061 8.75087 21.0976 8.80474 21.1464 8.85355C21.1953 8.90237 21.2491 8.93898 21.3081 8.96339C21.367 8.9878 21.431 9 21.5 9L27 9L28 10Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path></g></g>', 3)
    ]), 10, g2));
  }
}), L2 = ["width", "height", "aria-label"], m2 = /* @__PURE__ */ c({
  __name: "Mind",
  props: {
    size: { default: 32 },
    title: {},
    class: {}
  },
  setup(i) {
    const e = i;
    return (d, r) => (o(), a("svg", {
      width: e.size,
      height: e.size,
      viewBox: "0 0 32 32",
      xmlns: "http://www.w3.org/2000/svg",
      "xmlns:xlink": "http://www.w3.org/1999/xlink",
      fill: "none",
      customFrame: "#000000",
      role: "img",
      "aria-label": e.title,
      class: h(e.class)
    }, r[0] || (r[0] = [
      t(
        "mask",
        {
          id: "mask_12",
          width: "24",
          height: "28",
          x: "4",
          y: "2",
          maskUnits: "userSpaceOnUse",
          "mask-type": "outline"
        },
        [
          t("g", { filter: "url(#pixso_custom_mask_type_outline)" }, [
            t("use", { "xlink:href": "#use_12" })
          ])
        ],
        -1
        /* HOISTED */
      ),
      f('<defs><g id="use_12"><path id="蒙版" d="M7 2L20 2L28 10L28 27C28 28.6569 26.6569 30 25 30L7 30C5.34315 30 4 28.6569 4 27L4 5C4 3.34315 5.34315 2 7 2Z" fill="rgb(164,47,214)" fill-rule="evenodd"></path><path id="蒙版" d="M20 2L7 2C5.34315 2 4 3.34315 4 5L4 27C4 28.6569 5.34315 30 7 30L25 30C26.6569 30 28 28.6569 28 27L28 10L20 2ZM19.5858 3L27 10.4142L27 27C27 27.1577 26.9824 27.3113 26.9472 27.4607C26.9217 27.5687 26.8871 27.6745 26.8432 27.7782C26.7974 27.8867 26.7429 27.9896 26.6798 28.0869C26.6037 28.2041 26.5152 28.3132 26.4142 28.4142C26.3132 28.5152 26.2041 28.6037 26.0869 28.6797C25.9896 28.7429 25.8867 28.7974 25.7782 28.8432C25.6745 28.8871 25.5687 28.9217 25.4607 28.9472C25.3113 28.9824 25.1577 29 25 29L7 29C6.84229 29 6.68872 28.9824 6.53931 28.9472C6.43131 28.9217 6.32548 28.8871 6.22182 28.8432C6.11331 28.7973 6.01038 28.7428 5.91305 28.6797C5.79586 28.6037 5.68677 28.5152 5.58579 28.4142C5.48479 28.3132 5.39628 28.2041 5.32027 28.0869C5.25715 27.9896 5.20265 27.8867 5.15676 27.7782C5.11292 27.6745 5.07827 27.5687 5.05282 27.4607C5.01761 27.3113 5 27.1577 5 27L5 5C5 4.84228 5.01761 4.68871 5.05282 4.53929C5.07827 4.4313 5.11292 4.32547 5.15676 4.22182C5.20266 4.1133 5.25717 4.01037 5.32031 3.91303C5.39631 3.79585 5.4848 3.68677 5.58579 3.58579C5.68677 3.4848 5.79585 3.39631 5.91303 3.32031C6.01037 3.25717 6.1133 3.20266 6.22182 3.15676C6.32547 3.11292 6.4313 3.07827 6.53929 3.05282C6.68871 3.01761 6.84228 3 7 3L19.5858 3Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path></g><filter id="pixso_custom_mask_type_outline"><feColorMatrix type="matrix" values="0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 255 0 "></feColorMatrix></filter></defs><g id="思维导图"><rect id="矩形" width="32" height="32" x="0" y="0" fill="rgb(0,0,0)" fill-opacity="0"></rect><rect id="矩形" width="31" height="31" x="0.5" y="0.5" stroke="rgb(151,151,151)" stroke-opacity="0" stroke-width="1"></rect><g id="编组 2" mask="url(#mask_12)"><use xlink:href="#use_12"></use><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L28 10Z" fill="rgb(222,163,247)" fill-rule="evenodd"></path><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L21 3L21 8.5C21 8.56903 21.0122 8.63301 21.0366 8.69194C21.061 8.75087 21.0976 8.80474 21.1464 8.85355C21.1953 8.90237 21.2491 8.93898 21.3081 8.96339C21.367 8.9878 21.431 9 21.5 9L27 9L28 10Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path><g id="编组 33"><rect id="矩形" width="4" height="4" x="8.5" y="17" rx="2" fill="rgb(216,216,216)" fill-opacity="0"></rect><rect id="矩形" width="4" height="4" x="8.5" y="17" rx="2" stroke="rgb(255,255,255)" stroke-width="1.5"></rect><rect id="矩形备份 13" width="4" height="4" x="19.5" y="12.5" rx="2" fill="rgb(216,216,216)" fill-opacity="0"></rect><rect id="矩形备份 13" width="4" height="4" x="19.5" y="12.5" rx="2" stroke="rgb(255,255,255)" stroke-width="1.5"></rect><rect id="矩形备份 15" width="4" height="4" x="19.5" y="21.5" rx="2" fill="rgb(216,216,216)" fill-opacity="0"></rect><rect id="矩形备份 15" width="4" height="4" x="19.5" y="21.5" rx="2" stroke="rgb(255,255,255)" stroke-width="1.5"></rect><path id="路径 5" d="M12.5 19L16.0087 19" stroke="rgb(255,255,255)" stroke-width="1.5"></path><path id="路径" d="M19.446 23.5244L17.9496 23.5139C16.8505 23.5139 15.9636 22.6131 15.9636 21.514L15.9636 16.5141C15.9636 15.404 16.8676 14.5141 17.9776 14.5141L19.446 14.5244L19.446 23.5244Z" fill="rgb(216,216,216)" fill-opacity="0" fill-rule="evenodd"></path><path id="路径" d="M19.446 23.5244L17.9496 23.5139C16.8505 23.5139 15.9636 22.6131 15.9636 21.514L15.9636 16.5141C15.9636 15.404 16.8676 14.5141 17.9776 14.5141L19.446 14.5244" fill-rule="evenodd" stroke="rgb(255,255,255)" stroke-width="1.5"></path></g></g></g>', 2)
    ]), 10, L2));
  }
}), _2 = ["width", "height", "aria-label"], u = /* @__PURE__ */ c({
  __name: "Mp4",
  props: {
    size: { default: 32 },
    title: {},
    class: {}
  },
  setup(i) {
    const e = i;
    return (d, r) => (o(), a("svg", {
      width: e.size,
      height: e.size,
      viewBox: "0 0 32 32",
      xmlns: "http://www.w3.org/2000/svg",
      "xmlns:xlink": "http://www.w3.org/1999/xlink",
      fill: "none",
      customFrame: "#000000",
      role: "img",
      "aria-label": e.title,
      class: h(e.class)
    }, r[0] || (r[0] = [
      t(
        "mask",
        {
          id: "mask_9",
          width: "24",
          height: "28",
          x: "4",
          y: "2",
          maskUnits: "userSpaceOnUse",
          "mask-type": "outline"
        },
        [
          t("g", { filter: "url(#pixso_custom_mask_type_outline)" }, [
            t("use", { "xlink:href": "#use_9" })
          ])
        ],
        -1
        /* HOISTED */
      ),
      f('<defs><g id="use_9"><path id="蒙版" d="M7 2L20 2L28 10L28 27C28 28.6569 26.6569 30 25 30L7 30C5.34315 30 4 28.6569 4 27L4 5C4 3.34315 5.34315 2 7 2Z" fill="rgb(57,121,249)" fill-rule="evenodd"></path><path id="蒙版" d="M20 2L7 2C5.34315 2 4 3.34315 4 5L4 27C4 28.6569 5.34315 30 7 30L25 30C26.6569 30 28 28.6569 28 27L28 10L20 2ZM19.5858 3L27 10.4142L27 27C27 27.1577 26.9824 27.3113 26.9472 27.4607C26.9217 27.5687 26.8871 27.6745 26.8432 27.7782C26.7974 27.8867 26.7429 27.9896 26.6798 28.0869C26.6037 28.2041 26.5152 28.3132 26.4142 28.4142C26.3132 28.5152 26.2041 28.6037 26.0869 28.6797C25.9896 28.7429 25.8867 28.7974 25.7782 28.8432C25.6745 28.8871 25.5687 28.9217 25.4607 28.9472C25.3113 28.9824 25.1577 29 25 29L7 29C6.84229 29 6.68872 28.9824 6.53931 28.9472C6.43131 28.9217 6.32548 28.8871 6.22182 28.8432C6.11331 28.7973 6.01038 28.7428 5.91305 28.6797C5.79586 28.6037 5.68677 28.5152 5.58579 28.4142C5.48479 28.3132 5.39628 28.2041 5.32027 28.0869C5.25715 27.9896 5.20265 27.8867 5.15676 27.7782C5.11292 27.6745 5.07827 27.5687 5.05282 27.4607C5.01761 27.3113 5 27.1577 5 27L5 5C5 4.84228 5.01761 4.68871 5.05282 4.53929C5.07827 4.4313 5.11292 4.32547 5.15676 4.22182C5.20266 4.1133 5.25717 4.01037 5.32031 3.91303C5.39631 3.79585 5.4848 3.68677 5.58579 3.58579C5.68677 3.4848 5.79585 3.39631 5.91303 3.32031C6.01037 3.25717 6.1133 3.20266 6.22182 3.15676C6.32547 3.11292 6.4313 3.07827 6.53929 3.05282C6.68871 3.01761 6.84228 3 7 3L19.5858 3Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path></g><filter id="pixso_custom_mask_type_outline"><feColorMatrix type="matrix" values="0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 255 0 "></feColorMatrix></filter></defs><g id="mp4"><rect id="矩形" width="32" height="32" x="0" y="0" rx="8" fill="rgb(238,240,245)" fill-opacity="0"></rect><rect id="矩形" width="31" height="31" x="0.5" y="0.5" rx="8" stroke="rgb(151,151,151)" stroke-opacity="0" stroke-width="1"></rect><g id="路径" mask="url(#mask_9)"><use xlink:href="#use_9"></use><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L28 10Z" fill="rgb(157,189,252)" fill-rule="evenodd"></path><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L21 3L21 8.5C21 8.56903 21.0122 8.63301 21.0366 8.69194C21.061 8.75087 21.0976 8.80474 21.1464 8.85355C21.1953 8.90237 21.2491 8.93898 21.3081 8.96339C21.367 8.9878 21.431 9 21.5 9L27 9L28 10Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path><path id="矢量 780" d="M13.5 22.5433C13.5 22.6269 13.5965 22.6736 13.6621 22.6217L19.401 18.0784C19.4515 18.0384 19.4515 17.9616 19.401 17.9216L13.6621 13.3783C13.5965 13.3264 13.5 13.3731 13.5 13.4567L13.5 22.5433Z" stroke="rgb(255,255,255)" stroke-width="1.5"></path></g></g>', 2)
    ]), 10, _2));
  }
}), k2 = ["width", "height", "aria-label"], x2 = /* @__PURE__ */ c({
  __name: "Page",
  props: {
    size: { default: 32 },
    title: {},
    class: {}
  },
  setup(i) {
    const e = i;
    return (d, r) => (o(), a("svg", {
      width: e.size,
      height: e.size,
      viewBox: "0 0 32 32",
      xmlns: "http://www.w3.org/2000/svg",
      "xmlns:xlink": "http://www.w3.org/1999/xlink",
      fill: "none",
      customFrame: "#000000",
      role: "img",
      "aria-label": e.title,
      class: h(e.class)
    }, r[0] || (r[0] = [
      t(
        "mask",
        {
          id: "mask_7",
          width: "24",
          height: "28",
          x: "4",
          y: "2",
          maskUnits: "userSpaceOnUse",
          "mask-type": "outline"
        },
        [
          t("g", { filter: "url(#pixso_custom_mask_type_outline)" }, [
            t("use", { "xlink:href": "#use_7" })
          ])
        ],
        -1
        /* HOISTED */
      ),
      f('<defs><g id="use_7"><path id="蒙版" d="M7 2L20 2L28 10L28 27C28 28.6569 26.6569 30 25 30L7 30C5.34315 30 4 28.6569 4 27L4 5C4 3.34315 5.34315 2 7 2Z" fill="rgb(37,194,81)" fill-rule="evenodd"></path><path id="蒙版" d="M20 2L7 2C5.34315 2 4 3.34315 4 5L4 27C4 28.6569 5.34315 30 7 30L25 30C26.6569 30 28 28.6569 28 27L28 10L20 2ZM19.5858 3L27 10.4142L27 27C27 27.1577 26.9824 27.3113 26.9472 27.4607C26.9217 27.5687 26.8871 27.6745 26.8432 27.7782C26.7974 27.8867 26.7429 27.9896 26.6798 28.0869C26.6037 28.2041 26.5152 28.3132 26.4142 28.4142C26.3132 28.5152 26.2041 28.6037 26.0869 28.6797C25.9896 28.7429 25.8867 28.7974 25.7782 28.8432C25.6745 28.8871 25.5687 28.9217 25.4607 28.9472C25.3113 28.9824 25.1577 29 25 29L7 29C6.84229 29 6.68872 28.9824 6.53931 28.9472C6.43131 28.9217 6.32548 28.8871 6.22182 28.8432C6.11331 28.7973 6.01038 28.7428 5.91305 28.6797C5.79586 28.6037 5.68677 28.5152 5.58579 28.4142C5.48479 28.3132 5.39628 28.2041 5.32027 28.0869C5.25715 27.9896 5.20265 27.8867 5.15676 27.7782C5.11292 27.6745 5.07827 27.5687 5.05282 27.4607C5.01761 27.3113 5 27.1577 5 27L5 5C5 4.84228 5.01761 4.68871 5.05282 4.53929C5.07827 4.4313 5.11292 4.32547 5.15676 4.22182C5.20266 4.1133 5.25717 4.01037 5.32031 3.91303C5.39631 3.79585 5.4848 3.68677 5.58579 3.58579C5.68677 3.4848 5.79585 3.39631 5.91303 3.32031C6.01037 3.25717 6.1133 3.20266 6.22182 3.15676C6.32547 3.11292 6.4313 3.07827 6.53929 3.05282C6.68871 3.01761 6.84228 3 7 3L19.5858 3Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path></g><filter id="pixso_custom_mask_type_outline"><feColorMatrix type="matrix" values="0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 255 0 "></feColorMatrix></filter></defs><g id="Page"><g id="图标/word文件"><rect id="矩形" width="32" height="32" x="0" y="0" rx="4" fill="rgb(238,240,245)" fill-opacity="0"></rect><rect id="矩形" width="31" height="31" x="0.5" y="0.5" rx="4" stroke="rgb(151,151,151)" stroke-opacity="0" stroke-width="1"></rect><g id="编组 2" mask="url(#mask_7)"><use xlink:href="#use_7"></use><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L28 10Z" fill="rgb(146,224,165)" fill-rule="evenodd"></path><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L21 3L21 8.5C21 8.56903 21.0122 8.63301 21.0366 8.69194C21.061 8.75087 21.0976 8.80474 21.1464 8.85355C21.1953 8.90237 21.2491 8.93898 21.3081 8.96339C21.367 8.9878 21.431 9 21.5 9L27 9L28 10Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path><g id="编组 34"><path id="路径" d="M11.2618 22L10.5 22C9.39543 22 8.5 20.9553 8.5 19.6667L8.5 17.3333C8.5 16.0447 9.39543 15 10.5 15L11.4798 15L11.2618 22Z" fill="rgb(216,216,216)" fill-opacity="0" fill-rule="evenodd"></path><path id="路径" d="M11.2618 22L10.5 22C9.39543 22 8.5 20.9553 8.5 19.6667L8.5 17.3333C8.5 16.0447 9.39543 15 10.5 15L11.4798 15" fill-rule="evenodd" stroke="rgb(255,255,255)" stroke-width="1.5"></path><rect id="矩形" width="4" height="3" x="11.5" y="13.5" rx="1.5" fill="rgb(216,216,216)" fill-opacity="0"></rect><rect id="矩形" width="4" height="3" x="11.5" y="13.5" rx="1.5" stroke="rgb(255,255,255)" stroke-width="1.5"></rect><rect id="矩形备份 11" width="4" height="3" x="11.5" y="20.5" rx="1.5" fill="rgb(216,216,216)" fill-opacity="0"></rect><rect id="矩形备份 11" width="4" height="3" x="11.5" y="20.5" rx="1.5" stroke="rgb(255,255,255)" stroke-width="1.5"></rect><path id="矢量 774" d="M19 13.5L23 13.5" stroke="rgb(255,255,255)" stroke-linecap="round" stroke-width="1.5"></path><path id="矢量 775" d="M19 18.5L23 18.5" stroke="rgb(255,255,255)" stroke-linecap="round" stroke-width="1.5"></path><path id="矢量 776" d="M19 23.5L23 23.5" stroke="rgb(255,255,255)" stroke-linecap="round" stroke-width="1.5"></path></g></g></g></g>', 2)
    ]), 10, k2));
  }
}), w2 = ["width", "height", "aria-label"], D = /* @__PURE__ */ c({
  __name: "Pdf",
  props: {
    size: { default: 32 },
    title: {},
    class: {}
  },
  setup(i) {
    const e = i;
    return (d, r) => (o(), a("svg", {
      width: e.size,
      height: e.size,
      viewBox: "0 0 32 32",
      xmlns: "http://www.w3.org/2000/svg",
      "xmlns:xlink": "http://www.w3.org/1999/xlink",
      fill: "none",
      customFrame: "#000000",
      role: "img",
      "aria-label": e.title,
      class: h(e.class)
    }, r[0] || (r[0] = [
      t(
        "mask",
        {
          id: "mask_3",
          width: "24",
          height: "28",
          x: "4",
          y: "2",
          maskUnits: "userSpaceOnUse",
          "mask-type": "outline"
        },
        [
          t("g", { filter: "url(#pixso_custom_mask_type_outline)" }, [
            t("use", { "xlink:href": "#use_3" })
          ])
        ],
        -1
        /* HOISTED */
      ),
      f('<defs><g id="use_3"><path id="蒙版" d="M7 2L20 2L28 10L28 27C28 28.6569 26.6569 30 25 30L7 30C5.34315 30 4 28.6569 4 27L4 5C4 3.34315 5.34315 2 7 2Z" fill="rgb(242,48,48)" fill-rule="evenodd"></path><path id="蒙版" d="M20 2L7 2C5.34315 2 4 3.34315 4 5L4 27C4 28.6569 5.34315 30 7 30L25 30C26.6569 30 28 28.6569 28 27L28 10L20 2ZM19.5858 3L27 10.4142L27 27C27 27.1577 26.9824 27.3113 26.9472 27.4607C26.9217 27.5687 26.8871 27.6745 26.8432 27.7782C26.7974 27.8867 26.7429 27.9896 26.6798 28.0869C26.6037 28.2041 26.5152 28.3132 26.4142 28.4142C26.3132 28.5152 26.2041 28.6037 26.0869 28.6797C25.9896 28.7429 25.8867 28.7974 25.7782 28.8432C25.6745 28.8871 25.5687 28.9217 25.4607 28.9472C25.3113 28.9824 25.1577 29 25 29L7 29C6.84229 29 6.68872 28.9824 6.53931 28.9472C6.43131 28.9217 6.32548 28.8871 6.22182 28.8432C6.11331 28.7973 6.01038 28.7428 5.91305 28.6797C5.79586 28.6037 5.68677 28.5152 5.58579 28.4142C5.48479 28.3132 5.39628 28.2041 5.32027 28.0869C5.25715 27.9896 5.20265 27.8867 5.15676 27.7782C5.11292 27.6745 5.07827 27.5687 5.05282 27.4607C5.01761 27.3113 5 27.1577 5 27L5 5C5 4.84228 5.01761 4.68871 5.05282 4.53929C5.07827 4.4313 5.11292 4.32547 5.15676 4.22182C5.20266 4.1133 5.25717 4.01037 5.32031 3.91303C5.39631 3.79585 5.4848 3.68677 5.58579 3.58579C5.68677 3.4848 5.79585 3.39631 5.91303 3.32031C6.01037 3.25717 6.1133 3.20266 6.22182 3.15676C6.32547 3.11292 6.4313 3.07827 6.53929 3.05282C6.68871 3.01761 6.84228 3 7 3L19.5858 3Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path></g><filter id="pixso_custom_mask_type_outline"><feColorMatrix type="matrix" values="0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 255 0 "></feColorMatrix></filter></defs><g id="pdf"><g id="图标/pdf文件"><rect id="矩形" width="32" height="32" x="0" y="0" rx="8" fill="rgb(238,240,245)" fill-opacity="0"></rect><rect id="矩形" width="31" height="31" x="0.5" y="0.5" rx="8" stroke="rgb(151,151,151)" stroke-opacity="0" stroke-width="1"></rect><g id="编组 2"><g id="路径" mask="url(#mask_3)"><use xlink:href="#use_3"></use><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L28 10Z" fill="rgb(250,167,163)" fill-rule="evenodd"></path><path id="路径" d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L21 3L21 8.5C21 8.56903 21.0122 8.63301 21.0366 8.69194C21.061 8.75087 21.0976 8.80474 21.1464 8.85355C21.1953 8.90237 21.2491 8.93898 21.3081 8.96339C21.367 8.9878 21.431 9 21.5 9L27 9L28 10Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path><path id="path6" d="M16.364 18.6456C18.7712 21.8711 23.2379 23.4328 23.4902 21.3188C23.7425 19.2048 19.0713 19.4571 14.9933 20.5482C10.9085 21.6324 7.37606 24.1215 8.83541 25.5399C10.2879 26.9584 11.7814 24.7489 13.8067 20.0504C15.8321 15.345 17.5164 11.0352 15.1979 11.0352C12.8793 11.0352 13.9499 15.42 16.364 18.6456Z" fill="rgb(0,0,0)" fill-opacity="0" fill-rule="nonzero"></path><path id="path6" d="M23.4902 21.3188C23.7425 19.2048 19.0713 19.4571 14.9933 20.5482C10.9085 21.6324 7.37606 24.1215 8.83541 25.5399C10.2879 26.9584 11.7814 24.7489 13.8067 20.0504C15.8321 15.345 17.5164 11.0352 15.1979 11.0352C12.8793 11.0352 13.9499 15.42 16.364 18.6456C18.7712 21.8711 23.2379 23.4328 23.4902 21.3188Z" fill-rule="nonzero" stroke="rgb(255,255,255)" stroke-linejoin="round" stroke-width="1.5"></path></g></g></g></g>', 2)
    ]), 10, w2));
  }
}), y2 = ["width", "height", "aria-label"], G = /* @__PURE__ */ c({
  __name: "Ppt",
  props: {
    size: { default: 16 },
    title: {},
    class: {}
  },
  setup(i) {
    const e = i;
    return (d, r) => (o(), a("svg", {
      width: e.size,
      height: e.size,
      viewBox: "0 0 32 32",
      xmlns: "http://www.w3.org/2000/svg",
      role: "img",
      "aria-label": e.title,
      class: h(e.class)
    }, r[0] || (r[0] = [
      f('<g><!-- 橙色背景形状 --><path d="M7 2L20 2L28 10L28 27C28 28.6569 26.6569 30 25 30L7 30C5.34315 30 4 28.6569 4 27L4 5C4 3.34315 5.34315 2 7 2Z" fill="rgb(255,136,0)" fill-rule="evenodd"></path><!-- 内部白色 &quot;P&quot; 字母和装饰 --><g fill="#FFF" fill-rule="evenodd"><path d="M11.8164 9.26886L14.6559 9.26886C14.844 9.26886 14.9966 9.42141 14.9966 9.60959L14.9966 10.7454C14.9966 10.9336 14.844 11.0861 14.6559 11.0861L11.8164 11.0861L11.8164 9.26886Z"></path><path d="M11.8164 12.9036L14.6559 12.9036C14.844 12.9036 14.9966 13.0562 14.9966 13.2444L14.9966 14.7209L14.9966 14.7209L11.8164 14.7209L11.8164 12.9036Z"></path><path d="M0 0L2.83945 0C3.02763 0 3.18019 0.152552 3.18019 0.340734L3.18019 1.47652C3.18019 1.6647 3.02763 1.81725 2.83945 1.81725L0 1.81725L0 0Z" transform="matrix(-1,0,0,1,13.1802,7.45172)"></path><path d="M0.340734 0L3.18019 0L3.18019 0L3.18019 1.81725L0.340734 1.81725C0.152552 1.81725 0 1.6647 0 1.47652L0 0.340734C0 0.152552 0.152552 0 0.340734 0Z" transform="matrix(-1,0,0,1,14.9966,5.63434)"></path><path d="M0 0L3.18019 0L3.18019 1.81725L0.340734 1.81725C0.152552 1.81725 0 1.6647 0 1.47652L0 0L0 0Z" transform="matrix(-1,0,0,1,14.9966,2.00006)"></path><path d="M0 0L2.83945 0C3.02763 0 3.18019 0.152552 3.18019 0.340734L3.18019 1.47652C3.18019 1.6647 3.02763 1.81725 2.83945 1.81725L0 1.81725L0 0Z" transform="matrix(-1,0,0,1,13.1802,3.8172)"></path><path d="M0 0L2.83945 0C3.02763 0 3.18019 0.152552 3.18019 0.340734L3.18019 1.47652C3.18019 1.6647 3.02763 1.81725 2.83945 1.81725L0 1.81725L0 0Z" transform="matrix(-1,0,0,1,13.1802,11.0862)"></path><path d="M15 17.687L15 14.72L10.0026 14.72L10.0026 17.687C10.0026 18.0556 10.3015 18.3545 10.6701 18.3545L14.3324 18.3545C14.7011 18.3545 15 18.0556 15 17.687ZM11.2478 15.6313L13.7465 15.6313L13.7465 16.8807C13.7465 17.0689 13.5939 17.2214 13.4058 17.2214L11.5885 17.2214C11.4003 17.2214 11.2478 17.0689 11.2478 16.8807L11.2478 15.6313Z"></path></g><!-- 折角部分 --><path d="M28 10L21.5 10C20.6716 10 20 9.32843 20 8.5L20 2L28 10Z" fill="rgb(252,213,164)" fill-rule="evenodd"></path></g>', 1)
    ]), 10, y2));
  }
}), v2 = ["width", "height", "aria-label"], z = /* @__PURE__ */ c({
  __name: "Unknown",
  props: {
    size: { default: 32 },
    title: {},
    class: {}
  },
  setup(i) {
    const e = i;
    return (d, r) => (o(), a("svg", {
      width: e.size,
      height: e.size,
      viewBox: "0 0 33 35",
      xmlns: "http://www.w3.org/2000/svg",
      "xmlns:xlink": "http://www.w3.org/1999/xlink",
      fill: "none",
      customFrame: "#000000",
      role: "img",
      "aria-label": e.title,
      class: h(e.class)
    }, r[0] || (r[0] = [
      t(
        "mask",
        {
          id: "mask_15",
          width: "24",
          height: "28",
          x: "4",
          y: "5",
          maskUnits: "userSpaceOnUse",
          "mask-type": "outline"
        },
        [
          t("g", { filter: "url(#pixso_custom_mask_type_outline)" }, [
            t("use", { "xlink:href": "#use_15" })
          ])
        ],
        -1
        /* HOISTED */
      ),
      t(
        "defs",
        null,
        [
          t("g", { id: "use_15" }, [
            t("path", {
              id: "蒙版",
              d: "M7 5L22 5L28 11L28 30C28 31.6569 26.6569 33 25 33L7 33C5.34315 33 4 31.6569 4 30L4 8C4 6.34315 5.34315 5 7 5Z",
              fill: "rgb(189,189,189)",
              "fill-rule": "evenodd"
            }),
            t("path", {
              id: "蒙版",
              d: "M22 5L7 5C5.34315 5 4 6.34315 4 8L4 30C4 31.6569 5.34315 33 7 33L25 33C26.6569 33 28 31.6569 28 30L28 11L22 5ZM21.5858 6L27 11.4142L27 30C27 30.1577 26.9824 30.3113 26.9472 30.4607C26.9217 30.5687 26.8871 30.6745 26.8432 30.7782C26.7974 30.8867 26.7429 30.9896 26.6798 31.0869C26.6037 31.2041 26.5152 31.3132 26.4142 31.4142C26.3132 31.5152 26.2041 31.6037 26.0869 31.6797C25.9896 31.7429 25.8867 31.7974 25.7782 31.8432C25.6745 31.8871 25.5687 31.9217 25.4607 31.9472C25.3113 31.9824 25.1577 32 25 32L7 32C6.84229 32 6.68872 31.9824 6.53931 31.9472C6.43131 31.9217 6.32548 31.8871 6.22182 31.8432C6.11331 31.7973 6.01038 31.7428 5.91305 31.6797C5.79586 31.6037 5.68677 31.5152 5.58579 31.4142C5.48479 31.3132 5.39628 31.2041 5.32027 31.0869C5.25715 30.9896 5.20265 30.8867 5.15676 30.7782C5.11292 30.6745 5.07827 30.5687 5.05282 30.4607C5.01761 30.3113 5 30.1577 5 30L5 8C5 7.84228 5.01761 7.68871 5.05282 7.53929C5.07827 7.4313 5.11292 7.32547 5.15676 7.22182C5.20266 7.1133 5.25717 7.01037 5.32031 6.91303C5.39631 6.79585 5.4848 6.68677 5.58579 6.58579C5.68677 6.4848 5.79585 6.39631 5.91303 6.32031C6.01037 6.25717 6.1133 6.20266 6.22182 6.15676C6.32547 6.11292 6.4313 6.07827 6.53929 6.05282C6.68871 6.01761 6.84228 6 7 6L21.5858 6Z",
              fill: "rgb(223,225,230)",
              "fill-opacity": "0",
              "fill-rule": "evenodd"
            })
          ]),
          t("filter", { id: "pixso_custom_mask_type_outline" }, [
            t("feColorMatrix", {
              type: "matrix",
              values: "0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 255 0 "
            })
          ]),
          t("filter", {
            id: "filter_0",
            width: "16",
            height: "16",
            x: "17",
            y: "0",
            filterUnits: "userSpaceOnUse",
            customEffect: "url(#pixso_custom_effect_0)",
            "color-interpolation-filters": "sRGB"
          }, [
            t("feFlood", {
              "flood-opacity": "0",
              result: "BackgroundImageFix"
            }),
            t("feColorMatrix", {
              result: "hardAlpha",
              in: "SourceAlpha",
              type: "matrix",
              values: "0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0 "
            }),
            t("feOffset", {
              dx: "0",
              dy: "0"
            }),
            t("feGaussianBlur", { stdDeviation: "1.66666663" }),
            t("feComposite", {
              k2: "-1",
              k3: "1",
              in2: "hardAlpha",
              operator: "out"
            }),
            t("feColorMatrix", {
              type: "matrix",
              values: "0 0 0 0 0.643137 0 0 0 0 0.643137 0 0 0 0 0.643137 0 0 0 1 0 "
            }),
            t("feBlend", {
              result: "effect_dropShadow_1",
              in2: "BackgroundImageFix",
              mode: "normal"
            }),
            t("feBlend", {
              result: "shape",
              in: "SourceGraphic",
              in2: "effect_dropShadow_1",
              mode: "normal"
            })
          ])
        ],
        -1
        /* HOISTED */
      ),
      f('<g id="未知文件"><g id="图标/未知文件"><rect id="矩形" width="32" height="32" x="0" y="3" rx="8" fill="rgb(238,240,245)" fill-opacity="0"></rect><rect id="矩形" width="31" height="31" x="0.5" y="3.5" rx="8" stroke="rgb(151,151,151)" stroke-opacity="0" stroke-width="1"></rect><g id="路径" mask="url(#mask_15)"><use xlink:href="#use_15"></use><g id="ic_public_questionmark"><path id="path4" d="M15.25 26.22C15.25 25.8224 15.5724 25.5 15.97 25.5L16 25.5C16.4142 25.5 16.75 25.8358 16.75 26.25C16.75 26.6642 16.4142 27 16 27L15.97 27C15.5724 27 15.25 26.6776 15.25 26.28L15.25 26.28L15.25 26.22Z" fill="rgb(255,255,255)" fill-rule="evenodd"></path><path id="path4" d="M14.25 26.22C14.25 25.9872 14.2952 25.7639 14.3856 25.5502C14.4727 25.3443 14.5954 25.1622 14.7538 25.0038C14.9122 24.8454 15.0943 24.7227 15.3002 24.6356C15.5139 24.5452 15.7372 24.5 15.97 24.5L16 24.5C16.2369 24.5 16.4641 24.546 16.6815 24.6379C16.891 24.7265 17.0763 24.8514 17.2374 25.0126C17.3986 25.1737 17.5235 25.359 17.6121 25.5685C17.704 25.7859 17.75 26.0131 17.75 26.25C17.75 26.4869 17.704 26.7141 17.6121 26.9315C17.5235 27.141 17.3986 27.3263 17.2374 27.4874C17.0763 27.6486 16.891 27.7735 16.6815 27.8621C16.4641 27.954 16.2369 28 16 28L15.97 28C15.7372 28 15.5139 27.9548 15.3002 27.8644C15.0943 27.7773 14.9122 27.6546 14.7538 27.4962C14.5954 27.3378 14.4727 27.1557 14.3856 26.9498C14.2952 26.7361 14.25 26.5128 14.25 26.28L16.25 26.28C16.25 26.2027 16.2227 26.1367 16.168 26.082C16.1133 26.0273 16.0473 26 15.97 26L16 26C15.931 26 15.872 26.0244 15.8232 26.0732C15.7744 26.122 15.75 26.181 15.75 26.25C15.75 26.319 15.7744 26.378 15.8232 26.4268C15.872 26.4756 15.931 26.5 16 26.5L15.97 26.5C16.0473 26.5 16.1133 26.4727 16.168 26.418C16.2227 26.3633 16.25 26.2973 16.25 26.22L14.25 26.22ZM16.23 26.2L16.25 26.22C16.25 26.78 15.81 27.22 15.25 27.22C14.69 27.22 14.25 26.78 14.25 26.22L14.27 26.2L16.23 26.2ZMnan nanLnan nanCnan nan nan nan nan nanCnan nan nan nan nan nanLnan nan" fill="rgb(25,25,25)" fill-opacity="0" fill-rule="nonzero"></path><path id="path5" d="M16 23.5C16 23.2897 15.9663 22.5794 16 22.2727C16.028 22.0177 16.086 21.7987 16.2953 21.4564C16.6719 20.8404 17.836 20.1869 18.0114 20.0597C18.9117 19.407 19.5 18.3279 19.5 17.1072C19.5 15.115 17.933 13.5 16 13.5C14.067 13.5 12.5 15.115 12.5 17.1072" fill-rule="evenodd" stroke="rgb(255,255,255)" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5"></path></g><g filter="url(#filter_0)"><path id="路径" d="M28 11L24 11C22.8954 11 22 10.1046 22 9L22 5L28 11Z" fill="rgb(245,245,246)" fill-rule="evenodd"></path><path id="路径" d="M28 11L24 11C22.8954 11 22 10.1046 22 9L22 5L23 6L23 9C23 9.08264 23.0096 9.1629 23.0288 9.24079C23.0413 9.29117 23.0577 9.34055 23.0782 9.38894C23.1003 9.44134 23.1265 9.49116 23.1568 9.53839C23.1955 9.59887 23.2409 9.65511 23.2929 9.70711C23.3449 9.75911 23.4011 9.80448 23.4616 9.84322C23.5089 9.87346 23.5587 9.89966 23.6111 9.92183C23.6595 9.94229 23.7088 9.95874 23.7592 9.97117C23.8371 9.99039 23.9174 10 24 10L27 10L28 11Z" fill="rgb(223,225,230)" fill-opacity="0" fill-rule="evenodd"></path></g></g></g></g>', 1)
    ]), 10, v2));
  }
}), b2 = ["width", "height", "aria-hidden"], M2 = { key: 0 }, Z2 = /* @__PURE__ */ c({
  __name: "Wrong",
  props: {
    size: { default: 16 },
    title: {},
    class: {}
  },
  setup(i) {
    return (e, d) => (o(), a("svg", {
      xmlns: "http://www.w3.org/2000/svg",
      width: e.size,
      height: e.size,
      viewBox: "0 0 14 14",
      class: h(e.class),
      role: "img",
      "aria-hidden": !e.title,
      fill: "none"
    }, [
      e.title ? (o(), a(
        "title",
        M2,
        k(e.title),
        1
        /* TEXT */
      )) : C("v-if", !0),
      C(" 移除了不必要的 <g> 和 <clipPath> "),
      C(" 背景三角 "),
      d[0] || (d[0] = t(
        "path",
        {
          d: "M7.84 0.99L13.88 12.12C14.13 12.57 13.95 13.14 13.49 13.38C13.35 13.46 13.2 13.5 13.04 13.5H0.95C0.42 13.5 0 13.08 0 12.56C0 12.4 0.03 12.25 0.11 12.12L6.15 0.99C6.4 0.53 6.98 0.36 7.44 0.61C7.61 0.69 7.75 0.83 7.84 0.99Z",
          fill: "white"
        },
        null,
        -1
        /* HOISTED */
      )),
      C(" 感叹号的点 "),
      d[1] || (d[1] = t(
        "path",
        {
          d: "M7 11.5C6.58 11.5 6.25 11.16 6.25 10.75C6.25 10.33 6.58 10 7 10C7.41 10 7.75 10.33 7.75 10.75C7.75 11.16 7.41 11.5 7 11.5Z",
          fill: "#F23030"
        },
        null,
        -1
        /* HOISTED */
      )),
      C(" 感叹号的竖线 "),
      d[2] || (d[2] = t(
        "rect",
        {
          x: "6.5",
          y: "4.5",
          width: "1",
          height: "4.5",
          rx: "0.5",
          fill: "#F23030"
        },
        null,
        -1
        /* HOISTED */
      ))
    ], 10, b2));
  }
}), z2 = { class: "mc-file-list__container" }, F2 = ["onMouseenter"], $2 = { class: "mc-file-item__icon" }, U2 = ["src", "alt", "onClick"], B2 = {
  key: 2,
  class: "mc-file-item__progress-overlay"
}, S2 = {
  key: 1,
  class: "mc-file-item__progress-ring",
  viewBox: "0 0 36 36"
}, j2 = ["stroke-dashoffset"], O2 = { class: "mc-file-item__info" }, P2 = ["title"], E2 = { class: "mc-file-item__meta" }, R2 = ["title"], W2 = ["onClick"], D2 = ["onClick"], G2 = ["onClick"], I2 = { class: "mc-file-item__status" }, N2 = { class: "mc-file-item__file-type" }, q2 = { class: "mc-file-item__size" }, V2 = {
  key: 0,
  class: "mc-file-item__actions"
}, A2 = ["onClick", "title"], T2 = ["src", "alt"], H2 = ["src"], K2 = ["src"], J2 = { class: "mc-file-preview__unsupported" }, Q2 = { class: "mc-file-preview__unsupported-tip" }, X2 = ["title"], I = /* @__PURE__ */ c({
  name: "McFileList",
  __name: "FileList",
  props: r2,
  emits: s2,
  setup(i, { emit: e }) {
    const d = i, r = e, { t: L } = l2(), F = $(null), b = $(!1), _ = $(null), N = (s) => {
      if (s === 0) return "0 B";
      const n = 1024, l = ["B", "KB", "MB", "GB"], g = Math.floor(Math.log(s) / Math.log(n));
      return `${Number.parseFloat((s / n ** g).toFixed(1))} ${l[g]}`;
    }, q = (s) => (s.split(".").pop() || "File").toUpperCase(), S = {
      // Office & Documents
      ppt: G,
      pptx: G,
      pdf: D,
      doc: w,
      docx: w,
      xls: B,
      xlsx: B,
      csv: B,
      txt: w,
      rtf: w,
      page: x2,
      // Images
      jpg: y,
      jpeg: y,
      png: y,
      gif: y,
      bmp: y,
      webp: y,
      tif: y,
      tiff: y,
      // Vector & Design
      svg: Z,
      ai: Z,
      psd: Z,
      fig: Z,
      // Audio
      mp3: u,
      wav: u,
      ogg: u,
      flac: u,
      aac: u,
      m4a: u,
      // Videos
      mp4: u,
      mov: u,
      avi: u,
      mkv: u,
      webm: u,
      wmv: u,
      flv: u,
      // Archives
      zip: v,
      rar: v,
      "7z": v,
      tar: v,
      gz: v,
      bz2: v,
      // Code & Markup
      md: W,
      markdown: W,
      json: p,
      js: p,
      ts: p,
      jsx: p,
      tsx: p,
      html: p,
      css: p,
      scss: p,
      less: p,
      py: p,
      java: p,
      c: p,
      cpp: p,
      cs: p,
      go: p,
      php: p,
      rb: p,
      swift: p,
      kt: p,
      sh: p,
      xml: p,
      yml: p,
      yaml: p,
      sql: p,
      // Fonts
      ttf: w,
      otf: w,
      woff: w,
      woff2: w,
      // Others
      eml: C2,
      xmind: m2,
      flow: u2,
      // 'flowchart' is not a standard extension, maybe 'flow' is better
      dat: z,
      exe: z,
      dmg: z
    }, V = (s) => {
      var l, g, O, P, E, R;
      if (s.type) {
        if ((l = s.type) != null && l.startsWith("image/")) return y;
        if ((g = s.type) != null && g.startsWith("video/") || (O = s.type) != null && O.startsWith("audio/")) return u;
        if ((P = s.type) != null && P.startsWith("text/")) return w;
        if (s.type === "application/pdf") return D;
        if (s.type === "application/zip" || (E = s.type) != null && E.includes("compress"))
          return v;
      }
      const n = (R = s.name.split(".").pop()) == null ? void 0 : R.toLowerCase();
      return n && S[n] ? S[n] : z;
    }, A = (s) => {
      r("remove", s);
    }, T = (s) => {
      r("retry-upload", s);
    }, M = (s) => {
      var l;
      if (s.type) {
        if (s.type.startsWith("image/")) return "image";
        if (s.type.startsWith("video/")) return "video";
        if (s.type.startsWith("application/pdf") || s.type.startsWith("text/"))
          return "iframe";
      }
      const n = (l = s.name.split(".").pop()) == null ? void 0 : l.toLowerCase();
      if (n) {
        if ([
          "jpg",
          "jpeg",
          "png",
          "gif",
          "bmp",
          "webp",
          "svg",
          "tif",
          "tiff"
        ].includes(n))
          return "image";
        if (["mp4", "webm", "ogg"].includes(n)) return "video";
        if (["pdf", "txt"].includes(n)) return "iframe";
      }
      return "unsupported";
    }, j = (s) => {
      (s.url || s.thumbUrl) && (_.value = s, b.value = !0, r("preview", s));
    }, H = (s) => {
      r("retry-download", s);
    }, K = (s, n) => {
      r("download", s, n);
    };
    return (s, n) => (o(), a(
      "div",
      {
        class: h(["mc-file-list", `mc-file-list--context-${d.context}`])
      },
      [
        t("div", z2, [
          (o(!0), a(
            x,
            null,
            J(s.fileItems, (l) => (o(), a("div", {
              key: l.uid,
              class: h(["mc-file-item", [l.status ? `mc-file-item--${l.status}` : ""]]),
              onMouseenter: (g) => F.value = l.uid,
              onMouseleave: n[0] || (n[0] = (g) => F.value = null)
            }, [
              C(" 文件图标和进度 "),
              t("div", $2, [
                C(" 图片预览 "),
                M(l) === "image" && (l.thumbUrl || l.url) ? (o(), a("img", {
                  key: 0,
                  src: l.thumbUrl || l.url,
                  alt: l.name,
                  class: "mc-file-item__image-preview",
                  onClick: (g) => j(l)
                }, null, 8, U2)) : (o(), a(
                  x,
                  { key: 1 },
                  [
                    C(" 原来的图标 "),
                    (o(), U(Q(V(l)), {
                      title: l.name,
                      size: 36,
                      class: "mc-file-item__type-icon"
                    }, null, 8, ["title"]))
                  ],
                  64
                  /* STABLE_FRAGMENT */
                )),
                C(" 进度覆盖层 (同时处理上传和下载) "),
                l.status === "uploading" || l.status === "downloading" || l.status === "uploadError" || l.status === "downloadError" ? (o(), a("div", B2, [
                  n[4] || (n[4] = t(
                    "div",
                    { class: "mc-file-item__progress-mask" },
                    null,
                    -1
                    /* HOISTED */
                  )),
                  C(" 失败状态：显示 Wrong 图标 "),
                  l.status === "uploadError" || l.status === "downloadError" ? (o(), U(Z2, {
                    key: 0,
                    class: "mc-file-item__wrong-icon",
                    size: 12
                  })) : (o(), a("svg", S2, [
                    n[3] || (n[3] = t(
                      "path",
                      {
                        class: "mc-file-item__progress-ring-track",
                        d: "M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
                      },
                      null,
                      -1
                      /* HOISTED */
                    )),
                    t("path", {
                      class: "mc-file-item__progress-ring-circle",
                      "stroke-dasharray": "100, 100",
                      "stroke-dashoffset": 100 - (l.percentage || 0),
                      d: "M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
                    }, null, 8, j2)
                  ]))
                ])) : C("v-if", !0)
              ]),
              C(" 文件信息 "),
              t("div", O2, [
                t("div", {
                  class: "mc-file-item__name",
                  title: l.name
                }, k(l.name), 9, P2),
                t("div", E2, [
                  C(" 1. 失败状态 (统一处理上传和下载失败) "),
                  l.status === "uploadError" || l.status === "downloadError" ? (o(), a(
                    x,
                    { key: 0 },
                    [
                      t("span", {
                        class: "mc-file-item__status mc-file-item__status--error",
                        title: typeof l.error == "string" ? l.error : l.status === "uploadError" ? m(L)("FileList.uploadFailed") : m(L)("FileList.downloadFailed")
                      }, k(l.status === "uploadError" ? m(L)("FileList.uploadFailed") : m(L)("FileList.downloadFailed")), 9, R2),
                      t("span", {
                        class: "mc-file-item__meta-action",
                        onClick: (g) => l.status === "uploadError" ? T(l) : H(l)
                      }, k(m(L)("FileList.retry")), 9, W2)
                    ],
                    64
                    /* STABLE_FRAGMENT */
                  )) : F.value === l.uid && l.status === "success" ? (o(), a(
                    x,
                    { key: 1 },
                    [
                      C(" 2. 悬停状态 "),
                      t("span", {
                        class: "mc-file-item__meta-action",
                        onClick: (g) => K(l, g)
                      }, k(m(L)("FileList.download")), 9, D2),
                      t("span", {
                        class: "mc-file-item__meta-action",
                        onClick: (g) => j(l)
                      }, k(m(L)("FileList.preview")), 9, G2)
                    ],
                    64
                    /* STABLE_FRAGMENT */
                  )) : l.status === "uploading" || l.status === "downloading" ? (o(), a(
                    x,
                    { key: 2 },
                    [
                      C(" 3. 上传/下载中状态 "),
                      t(
                        "span",
                        I2,
                        k(m(L)(`FileList.${l.status}`)),
                        1
                        /* TEXT */
                      )
                    ],
                    64
                    /* STABLE_FRAGMENT */
                  )) : (o(), a(
                    x,
                    { key: 3 },
                    [
                      C(" 4. 默认状态 "),
                      t(
                        "span",
                        N2,
                        k(q(l.name)),
                        1
                        /* TEXT */
                      ),
                      t(
                        "span",
                        q2,
                        k(N(l.size)),
                        1
                        /* TEXT */
                      )
                    ],
                    64
                    /* STABLE_FRAGMENT */
                  ))
                ])
              ]),
              C(" 删除按钮 "),
              d.context === "input" ? (o(), a("div", V2, [
                t("button", {
                  class: "mc-file-item__action-btn mc-file-item__action-btn--remove",
                  onClick: (g) => A(l),
                  title: m(L)("FileList.remove")
                }, " ✕ ", 8, A2)
              ])) : C("v-if", !0)
            ], 42, F2))),
            128
            /* KEYED_FRAGMENT */
          ))
        ]),
        (o(), U(X, { to: "body" }, [
          Y(t2, { name: "mc-file-preview-fade" }, {
            default: e2(() => [
              b.value && _.value ? (o(), a("div", {
                key: 0,
                class: "mc-file-preview__overlay",
                onClick: n[2] || (n[2] = i2((l) => b.value = !1, ["self"]))
              }, [
                C(" 图片预览 "),
                M(_.value) === "image" ? (o(), a("img", {
                  key: 0,
                  src: _.value.thumbUrl || _.value.url,
                  alt: _.value.name,
                  class: "mc-file-preview__content"
                }, null, 8, T2)) : M(_.value) === "video" ? (o(), a(
                  x,
                  { key: 1 },
                  [
                    C(" 视频预览 "),
                    t("video", {
                      src: _.value.url,
                      controls: "",
                      class: "mc-file-preview__content"
                    }, null, 8, H2)
                  ],
                  2112
                  /* STABLE_FRAGMENT, DEV_ROOT_FRAGMENT */
                )) : M(_.value) === "iframe" ? (o(), a(
                  x,
                  { key: 2 },
                  [
                    C(" PDF 和 文本文件预览 (使用 iframe) "),
                    t("iframe", {
                      src: _.value.url,
                      class: "mc-file-preview__content mc-file-preview__iframe"
                    }, null, 8, K2)
                  ],
                  2112
                  /* STABLE_FRAGMENT, DEV_ROOT_FRAGMENT */
                )) : (o(), a(
                  x,
                  { key: 3 },
                  [
                    C(" 其他文件类型的占位符 "),
                    t("div", J2, [
                      t(
                        "span",
                        null,
                        k(m(L)("FileList.unsupportedPreview", { fileName: _.value.name })),
                        1
                        /* TEXT */
                      ),
                      t(
                        "span",
                        Q2,
                        k(m(L)("FileList.tryDownload")),
                        1
                        /* TEXT */
                      )
                    ])
                  ],
                  2112
                  /* STABLE_FRAGMENT, DEV_ROOT_FRAGMENT */
                )),
                C(" 关闭按钮 "),
                t("button", {
                  class: "mc-file-preview__close-btn",
                  onClick: n[1] || (n[1] = (l) => b.value = !1),
                  title: m(L)("FileList.close")
                }, "✕", 8, X2)
              ])) : C("v-if", !0)
            ]),
            _: 1
            /* STABLE */
          })
        ]))
      ],
      2
      /* CLASS */
    ));
  }
});
I.install = (i) => {
  i.component("McFileList", I);
};
export {
  I as McFileList
};
